// @webex/webex-sign-in-page/src/WebexMicrosoftSignInPage.tsx
import { ButtonPill } from "@momentum-ui/react-collaboration";
import React, { ReactElement } from "react";
import { useTranslation } from 'react-i18next';
import './WebexMicrosoftSignInPage.scss';


type IWebexMicrosoftSignInProps = {
  signInImage: ReactElement;
  signInTitle: string;
  signInMessage: string;
  signInButtonText: string;
  footNote: ReactElement;
  showSignInOrRetry: boolean;
  signInAction?: () => void;
  signingIn: boolean;
  clientType : string;
};

export const WebexMicrosoftSignInPage = ({
  signInImage,
  signInTitle,
  signInMessage,
  signInButtonText,
  footNote,
  showSignInOrRetry,
  signInAction,
  signingIn,
  clientType
  }: IWebexMicrosoftSignInProps) => {
    const { t } = useTranslation('WebexSpeedDials');
    return (
      <div className={ `sign-in-page ${ clientType !== 'desktop' && clientType!== 'web' ? 'sign-in-page__mobile': '' }`}>
        {!signingIn && (
        <div className={`webex-microsoft-message-container ${clientType!=='desktop' && clientType!=='web'? 'webex-microsoft-message-container__mobile':''} ${ showSignInOrRetry ? `webex-microsoft-message-container__error` : ''}`}>
          <div className="icon">
            { signInImage }
          </div>
          <div className="signInText">
            <p className="title">{signInTitle}</p>
            <p className="message">{signInMessage}</p>
          </div>
          <div className="close">
            <ButtonPill onPress={signInAction}>
              { signInButtonText }
            </ButtonPill>
          </div>
          { footNote }
        </div>
        )}
        {signingIn && (
        <div>{t('signIn.signingIn')}</div>
        )}
      </div>
    );
  };
