// @ciscospark/widget-roster/src/actions.js
export const VIEW_MAIN = `VIEW_MAIN`;
export const VIEW_ADD = `VIEW_ADD`;

export const SET_PARTICIPANT_EDITING = `widget-roster/SET_PARTICIPANT_EDITING`;
export const UPDATE_WIDGET_STATE = `widget-roster/UPDATE_WIDGET_STATE`;

/**
 * Sets if a participant is being edited
 *
 * @export
 * @param {object|boolean} participant
 * @returns {Object}
 */
export function setParticipantEditing(participant) {
  return {
    type: SET_PARTICIPANT_EDITING,
    payload: {
      participant
    }
  };
}

export function updateWidgetState(state) {
  return {
    type: UPDATE_WIDGET_STATE,
    payload: {
      state
    }
  };
}

/**
 * Sets the widget view to the main roster view
 * @returns {Object}
 */
export function setWidgetViewMain() {
  return updateWidgetState({currentView: VIEW_MAIN});
}

/**
 * Sets the widget view to the add participant view
 * @returns {Object}
 */
export function setWidgetViewAdd() {
  return updateWidgetState({currentView: VIEW_ADD});
}

/**
 * Sets the current search term
 * @param {string} searchTerm
 * @returns {Object}
 */
export function setSearchTerm(searchTerm) {
  return updateWidgetState({searchTerm});
}
