// @ciscospark/widget-space/src/reducer.js
import {fromJS} from 'immutable';
import {reducers as message} from '@ciscospark/widget-message';
import {reducers as meet} from '@ciscospark/widget-meet';
import mercury from '@ciscospark/redux-module-mercury';
import media from '@ciscospark/redux-module-media';
import user from '@ciscospark/redux-module-user';

import {
  FETCHING_SPACE_DETAILS,
  STORE_SPACE_DETAILS,
  STORE_ACTIVITY_TYPES,
  TOGGLE_ACTIVITY_MENU_VISIBLE,
  UPDATE_ACTIVITY_MENU_VISIBLE,
  UPDATE_ACTIVITY_TYPE
} from './actions';

export const initialState = fromJS({
  activityType: `message`,
  activityTypes: [],
  spaceDetails: null,
  status: {
    activityMenuVisible: false,
    fetchingSpaceDetails: false
  }
});

export function reducer(state = initialState, action) {
  switch (action.type) {
  case UPDATE_ACTIVITY_MENU_VISIBLE:
    return state.set(`activityMenuVisible`, action.payload.isActivityMenuVisible);

  case UPDATE_ACTIVITY_TYPE:
    return state.set(`activityType`, action.payload.type)
      .setIn([`status`, `activityMenuVisible`], false);

  case TOGGLE_ACTIVITY_MENU_VISIBLE:
    return state.setIn([`status`, `activityMenuVisible`], !state.getIn([`status`, `activityMenuVisible`]));

  case FETCHING_SPACE_DETAILS:
    return state.setIn([`status`, `fetchingSpaceDetails`], true);

  case STORE_SPACE_DETAILS:
    return state.setIn([`status`, `fetchingSpaceDetails`], false)
      .set(`spaceDetails`, fromJS(action.payload.details));

  case STORE_ACTIVITY_TYPES: {
    const activityTypes = state.get(`activityTypes`).concat(fromJS(action.payload.activityTypes));
    return state.set(`activityTypes`, activityTypes);
  }

  default:
    return state;
  }
}

const reducers = {
  ...message,
  ...meet,
  user,
  mercury,
  media,
  widgetSpace: reducer
};

export default reducers;
