// @ciscospark/react-component-error-display/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import styles from './styles.css';

export default function ErrorDisplay(props) {
  const {
    actionTitle,
    onAction,
    secondaryTitle,
    title,
    transparent
  } = props;

  function handleAction(e) {
    e.preventDefault();
    onAction();
  }

  return (
    <div className={classNames('ciscospark-error-display', styles.errorDisplay, {[styles.transparent]: transparent})}>
      <div className={classNames('ciscospark-error-content', styles.content)}>
        <div className={classNames('ciscospark-error-title', styles.errorTitle, styles.contentItem)}>
          {title}
        </div>
        {
          secondaryTitle
          && <div className={classNames('ciscospark-error-title-secondary', styles.secondaryTitle, styles.contentItem)}>
            {secondaryTitle}
             </div>
        }
        {
          actionTitle && onAction
          && <button
            className={classNames('ciscospark-link-button', styles.linkButton)}
            onClick={handleAction}
            onKeyPress={handleAction}
          >
            {actionTitle}
             </button>
        }
      </div>
    </div>
  );
}

ErrorDisplay.propTypes = {
  actionTitle: PropTypes.string,
  onAction: PropTypes.func,
  secondaryTitle: PropTypes.string,
  title: PropTypes.string.isRequired,
  transparent: PropTypes.bool
};
