// @webex/WebexSignInPage/WebexSignInPage.tsx
import { Button } from "@momentum-ui/react";
import React, { ReactElement } from "react";
import './WebexSignInPage.scss';

type IWebexSignInProps = {
  signInImage: ReactElement;
  signInTitle: string;
  signInMessage: string;
  signInButtonText: string;
  footNote: ReactElement;
  showSignInOrRetry: boolean;
  signInAction?: () => void;
};

export const WebexSignInPage = ({
  signInImage,
  signInTitle,
  signInMessage,
  signInButtonText,
  footNote,
  showSignInOrRetry,
  signInAction,
  }: IWebexSignInProps) => {
    return (
      <div className='sign-in-page'>   
        <div className={`webex-message-container ${ showSignInOrRetry ? `webex-message-container__error` : ''}`}>
          <div className="icon">
            { signInImage }
          </div>
          <p className="title">{signInTitle}</p>
          <p className="message">{signInMessage}</p>
          <div className="close">
            <Button onPress={signInAction}>
              { signInButtonText }
            </Button>
          </div>
          { footNote }
        </div>
      </div>
    );
  };
