// @ciscospark/react-redux-spark/src/inject-spark.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';

import SparkComponent from './component';

function getDisplayName(C) {
  return C.displayName || C.name || 'C';
}

export default function injectSpark(WrappedComponent, options = {}) {
  const {
    withRef = false
  } = options;

  class InjectSpark extends Component {
    shouldComponentUpdate(nextProps) {
      return nextProps !== this.props;
    }

    getWrappedInstance() {
      // eslint-disable-next-line react/no-string-refs
      return this.refs.wrappedInstance;
    }

    render() {
      const {accessToken} = this.props;
      return (
        <div>
          {
            accessToken
            && <SparkComponent accessToken={accessToken} {...this.props} />
          }
          <WrappedComponent
            {...this.props}
            ref={withRef ? 'wrappedInstance' : null}
          />
        </div>
      );
    }
  }

  InjectSpark.propTypes = {
    accessToken: PropTypes.string
  };

  InjectSpark.displayName = `InjectSpark(${getDisplayName(WrappedComponent)})`;
  InjectSpark.WrappedComponent = WrappedComponent;

  return InjectSpark;
}
