// @ciscospark/react-redux-spark/src/spark.js
/* eslint camelcase: [0] */

import '@ciscospark/plugin-logger';
import '@ciscospark/plugin-mercury';
import '@ciscospark/plugin-people';
import '@ciscospark/plugin-conversation';
import '@ciscospark/plugin-phone';
import '@ciscospark/plugin-metrics';
import '@ciscospark/plugin-flag';
import Spark from '@ciscospark/spark-core';
import LocalStorageStoreAdapter from '@ciscospark/storage-adapter-local-storage';
import LocalForageStoreAdapter from '@ciscospark/storage-adapter-local-forage';

export default function createSpark(accessToken) {
  // Check if we already have a spark instance in memory
  let ciscosparkInstance = window.ciscosparkInstance;

  if (typeof ciscosparkInstance !== `object`) {
    const credentials = JSON.parse(localStorage.getItem(`credentials`) || false);
    ciscosparkInstance = new Spark({
      credentials: {
        authorization: {
          access_token: accessToken || credentials.access_token
        }
      },
      config: {
        logger: {
          level: process.env.NODE_ENV === `production` ? `silent` : `trace`
        },
        conversation: {
          allowedInboundTags: {
            'spark-mention': [`data-object-type`, `data-object-id`, `data-object-url`],
            a: [`href`],
            b: [],
            blockquote: [`class`],
            strong: [],
            i: [],
            em: [],
            pre: [],
            code: [],
            br: [],
            hr: [],
            p: [],
            ul: [],
            ol: [],
            li: [],
            h1: [],
            h2: [],
            h3: []
          },
          allowedOutboundTags: {
            'spark-mention': [`data-object-type`, `data-object-id`, `data-object-url`],
            a: [`href`],
            b: [],
            blockquote: [`class`],
            strong: [],
            i: [],
            em: [],
            pre: [],
            code: [],
            br: [],
            hr: [],
            p: [],
            ul: [],
            ol: [],
            li: [],
            h1: [],
            h2: [],
            h3: []
          }
        },
        storage: {
          boundedAdapter: new LocalStorageStoreAdapter(`ciscospark-embedded`),
          unboundedAdapter: new LocalForageStoreAdapter(`ciscospark-widgets`)
        }
      }
    });
    window.ciscosparkInstance = ciscosparkInstance;
  }

  return ciscosparkInstance;
}
