// @ciscospark/react-component-confirmation-modal/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import styles from './styles.css';

export default function ConfirmationModal(props) {
  const {
    title,
    actionButtonLabel,
    cancelButtonLabel,
    body
  } = props;

  const modalClassNames = classNames(`ciscospark-dialogue-modal`, styles.dialogueModal);

  const actionBtnClassNames = classNames(
    `dialogue-modal-btn`,
    `dialogue-modal-action-btn`,
    styles.dialogueModalBtn,
    styles.dialogueModalActionBtn
  );
  const cancelBtnClassNames = classNames(
    `dialogue-modal-btn`,
    `dialogue-modal-exit-btn`,
    styles.dialogueModalBtn,
    styles.dialogueModalExitBtn
  );

  return (
    <div>
      <div className={modalClassNames}>
        <div className={classNames(`ciscospark-dialogue-modal-body`, styles.dialogModalBody)}>
          <div className={classNames(`ciscospark-dialogue-modal-title-text`, styles.dialogueModalTitleText)}>
            {title}
          </div>
          <div className={classNames(`ciscospark-dialogue-modal-body-subtext`, styles.dialogueModalBodySubtext)}>
            {body}
          </div>
          <button
            className={actionBtnClassNames}
            onClick={props.onClickActionButton}
            title={actionButtonLabel}
          >
            {actionButtonLabel}
          </button>
          <button
            className={cancelBtnClassNames}
            onClick={props.onClickCancelButton}
            title={cancelButtonLabel}
          >
            {cancelButtonLabel}
          </button>
        </div>
      </div>
      <div className={classNames(`ciscospark-dialogue-modal-backdrop`, styles.dialogueModalBackdrop)} />
    </div>
  );

}

ConfirmationModal.propTypes = {
  actionButtonLabel: PropTypes.string,
  body: PropTypes.string,
  cancelButtonLabel: PropTypes.string,
  onClickActionButton: PropTypes.func,
  onClickCancelButton: PropTypes.func,
  title: PropTypes.string
};
