// @ciscospark/react-component-activity-list/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import ActivityItem from '@ciscospark/react-component-activity-item';
import DaySeparator from '@ciscospark/react-component-day-separator';
import NewMessagesSeparator from '@ciscospark/react-component-new-messages-separator';
import {formatDate} from '@ciscospark/react-component-utils';

import styles from './styles.css';

export const ITEM_TYPE_ACTIVITY = `ITEM_TYPE_ACITIVITY`;
export const ITEM_TYPE_DAY_SEPARATOR = `ITEM_TYPE_DAY_SEPARATOR`;
export const ITEM_TYPE_NEW_MESSAGE_SEPARATOR = `ITEM_TYPE_NEW_MESSAGE_SEPARATOR`;

export default function ActivityList(props) {
  const {activities, onActivityDelete, onActivityFlag} = props;
  const items = activities.map((visibleActivity) => {
    switch (visibleActivity.type) {
    case ITEM_TYPE_DAY_SEPARATOR: {
      const {fromDate, key, now, toDate} = visibleActivity;
      return (
        <div className={classNames(styles.separator, `activity-list-separator`)}>
          <DaySeparator
            fromDate={fromDate}
            key={key}
            now={now}
            toDate={toDate}
          />
        </div>
      );
    }
    case ITEM_TYPE_ACTIVITY: {
      const {activity, avatarUrl, isPending, isAdditional, isFlagged, isSelf} = visibleActivity;

      return (
        <ActivityItem
          activity={activity.object}
          avatarUrl={avatarUrl}
          id={activity.id}
          isAdditional={isAdditional}
          isFlagged={isFlagged}
          isPending={isPending}
          isSelf={isSelf}
          key={activity.id}
          name={activity.actor.displayName}
          onActivityDelete={onActivityDelete}
          onActivityFlag={onActivityFlag}
          timestamp={formatDate(activity.published)}
          verb={activity.verb}
        />
      );
    }
    case ITEM_TYPE_NEW_MESSAGE_SEPARATOR: {
      return <NewMessagesSeparator key={visibleActivity.key} />;
    }
    default: {
      return ``;
    }
    }
  });

  return (
    <div className={classNames(`activity-list`)}>
      {items}
    </div>
  );
}

ActivityList.propTypes = {
  activities: PropTypes.arrayOf(PropTypes.shape({
    activity: PropTypes.shape({
      id: PropTypes.string,
      object: PropTypes.object,
      actor: PropTypes.object,
      published: PropTypes.string,
      verb: PropTypes.string
    }),
    type: PropTypes.string,
    avatarUrl: PropTypes.string,
    isAdditional: PropTypes.bool,
    isFlagged: PropTypes.bool,
    isSelf: PropTypes.bool,
    isPending: PropTypes.bool
  })),
  onActivityDelete: PropTypes.func.isRequired,
  onActivityFlag: PropTypes.func.isRequired
};
