// @ciscospark/widget-message-meet/src/actions.js
export const TOGGLE_ACTIVITY_MENU = `TOGGLE_ACTIVITY_MENU`;
export function toggleActivityMenu() {
  return {
    type: TOGGLE_ACTIVITY_MENU
  };
}

export const ACTIVITY_TYPE_MESSAGE = `ACTIVITY_TYPE_MESSAGE`;
export const ACTIVITY_TYPE_MEET = `ACTIVITY_TYPE_MEET`;
export const CHANGE_ACTIVITY_TYPE = `CHANGE_ACTIVITY_TYPE`;
function changeActivityType(type) {
  return {
    type: CHANGE_ACTIVITY_TYPE,
    payload: {
      type
    }
  };
}

export const SHOW_ACTIVITY_MENU = `SHOW_ACTIVITY_MENU`;
export function showActivityMenu() {
  return {
    type: SHOW_ACTIVITY_MENU
  };
}

/**
 * Sets the primary activity to a message
 *
 * @export
 * @returns {Promise}
 */
export function setActivityMessage() {
  return (dispatch) => {
    dispatch(changeActivityType(ACTIVITY_TYPE_MESSAGE));
  };
}

/**
 * Sets the primary activity to a meet (video)
 *
 * @export
 * @returns {Promise}
 */
export function setActivityMeet() {
  return (dispatch) => {
    dispatch(changeActivityType(ACTIVITY_TYPE_MEET));
  };
}
