// @ciscospark/widget-message-meet/src/container.js
import React, {Component, PropTypes} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import classNames from 'classnames';
import autobind from 'autobind-decorator';

import {fetchCurrentUser, fetchToUser} from '@ciscospark/redux-module-user';
import {injectSpark} from '@ciscospark/react-redux-spark';
import TitleBar from '@ciscospark/react-component-title-bar';
import Notifications from '@ciscospark/react-container-notifications';
import ActivityMenu from './components/activity-menu';
import LoadingScreen from '@ciscospark/react-component-loading-screen';
import {ConnectedMessageComponent} from '@ciscospark/widget-message';
import {ConnectedMeetComponent} from '@ciscospark/widget-meet';
import Button from '@ciscospark/react-component-button';
import {ICON_TYPE_WAFFLE} from '@ciscospark/react-component-icon';
import IncomingCall from '@ciscospark/react-container-call-incoming';

import {
  toggleActivityMenu,
  setActivityMeet,
  setActivityMessage,
  ACTIVITY_TYPE_MEET,
  ACTIVITY_TYPE_MESSAGE
} from './actions';
import styles from './styles.css';

/**
 * MessageMeetWidget Component
 */
export class MessageMeetWidget extends Component {

  componentWillReceiveProps(nextProps) {
    const {
      media,
      user,
      spark,
      sparkState,
      toPersonEmail,
      toPersonId,
      widgetMessageMeet
    } = nextProps;

    const {
      authenticated,
      connected,
      registered
    } = sparkState;

    if (spark && connected && authenticated && registered) {
      if (!user.get(`currentUser`) && !user.getIn([`status`, `isFetchingCurrentUser`])) {
        nextProps.fetchCurrentUser(spark);
      }
      if (!user.get(`toUser`) && !user.getIn([`status`, `isFetchingToUser`])) {
        nextProps.fetchToUser({toPersonEmail, toPersonId}, spark);
      }
      // When receivng an incoming call, switch to meet
      if (media.getIn([`status`, `isRinging`]) && widgetMessageMeet.get(`activityType`) !== ACTIVITY_TYPE_MEET) {
        nextProps.setActivityMeet();
      }
    }
  }

  shouldComponentUpdate(nextProps) {
    const props = this.props;

    return nextProps.sparkState.connected !== props.sparkState.connected
      || nextProps.user !== props.user
      || nextProps.conversation !== props.conversation
      || nextProps.avatar !== props.avatar
      || nextProps.indicators !== props.indicators
      || nextProps.widgetMessageMeet !== props.widgetMessageMeet;
  }

  @autobind
  handleActivityMeet() {
    const props = this.props;
    props.setActivityMeet();
    props.toggleActivityMenu();
  }

  @autobind
  handleActivityMessage() {
    const props = this.props;
    props.setActivityMessage();
    props.toggleActivityMenu();
  }

  @autobind
  handleMenuClick() {
    const props = this.props;
    props.toggleActivityMenu();
  }

  @autobind
  handleEvent(name, data) {
    const {onEvent} = this.props;
    if (typeof onEvent === `function`) {
      this.props.onEvent(name, data);
    }
  }

  /**
   * Render
   *
   * @returns {Object}
   */
  render() {
    const props = this.props;
    const {
      avatar,
      conversation,
      spark,
      user,
      widgetMessageMeet
    } = props;

    const toUser = user.get(`toUser`);
    const showActivityMenu = widgetMessageMeet.get(`showActivityMenu`);
    const showMessage = widgetMessageMeet.get(`activityType`) === ACTIVITY_TYPE_MESSAGE;
    const showMeet = widgetMessageMeet.get(`activityType`) === ACTIVITY_TYPE_MEET;
    // Don't remove message component from DOM, just hide its display
    const messageComponentClassNames = classNames(`message-component-wrapper`, styles.messageComponentWrapper, showMessage ? `` : styles.hiddenComponent);

    // Don't load widgets until we have spark loaded and info on the toUser
    if (spark && toUser && toUser.id && conversation) {
      const toUserAvatarUrl = avatar.getIn([`items`, toUser.id]);
      return (
        <div className={classNames(`widget-message-meet`, styles.widgetMessageMeet)}>
          <div className={classNames(`banner`, styles.banner)} />
          <div className={classNames(`widget-message-meet-inner`, styles.widgetMessageMeetInner)}>
            {
              showActivityMenu &&
              <div className={classNames(`activity-menu-wrapper`, styles.activityMenuWrapper)}>
                <ActivityMenu
                  onExit={this.handleMenuClick}
                  onMeet={this.handleActivityMeet}
                  onMessage={this.handleActivityMessage}
                />
              </div>
            }
            <div className={classNames(`activity-menu-button-wrapper`, styles.activityMenuButtonWrapper)}>
              <div className={classNames(`activity-menu-button`, styles.activityMenuButton)}>
                <Button iconType={ICON_TYPE_WAFFLE} onClick={this.handleMenuClick} />
              </div>
            </div>
            <div className={classNames(`title-bar-wrapper`, styles.titleBarWrapper)}>
              <TitleBar
                hasMenu
                image={toUserAvatarUrl}
                name={toUser.displayName}
                onMenuClick={this.handleMenuClick}
              />
            </div>
            {
              showMeet &&
              <ConnectedMeetComponent
                onEvent={this.handleEvent}
                spark={spark}
                toPersonAvatar={toUserAvatarUrl}
                toPersonId={toUser.id}
                toPersonName={toUser.displayName}
                user={user}
              />
            }
            <div className={messageComponentClassNames}>
              <ConnectedMessageComponent
                onEvent={this.handleEvent}
                spark={spark}
                toPersonId={toUser.id}
                user={user}
              />
            </div>
            <Notifications />
            <IncomingCall onEvent={this.handleEvent} />
          </div>
        </div>
      );
    }
    // Show loading screen until spark is ready
    return <LoadingScreen />;
  }
}

const injectedPropTypes = {
  conversation: PropTypes.object,
  fetchCurrentUser: PropTypes.func,
  fetchToUser: PropTypes.func,
  media: PropTypes.object,
  spark: PropTypes.object,
  sparkState: PropTypes.object,
  user: PropTypes.object
};

MessageMeetWidget.propTypes = {
  initialActivity: PropTypes.string, // eslint-disable-line react/no-unused-prop-types
  onEvent: PropTypes.func,
  startCall: PropTypes.bool, // eslint-disable-line react/no-unused-prop-types
  toPersonEmail: PropTypes.string,
  toPersonId: PropTypes.string,
  ...injectedPropTypes
};

function mapStateToProps(state) {
  return {
    avatar: state.avatar,
    conversation: state.conversation,
    media: state.media,
    spark: state.spark.get(`spark`),
    sparkState: state.spark.get(`status`).toJS(),
    user: state.user,
    indicators: state.indicators,
    widgetMessageMeet: state.widgetMessageMeet
  };
}

export default connect(
  mapStateToProps,
  (dispatch) => bindActionCreators({
    fetchCurrentUser,
    fetchToUser,
    setActivityMeet,
    setActivityMessage,
    toggleActivityMenu
  }, dispatch)
)(injectSpark(MessageMeetWidget));
