// @ciscospark/react-container-activity-list/src/formatters/at-mention.js
import React from 'react';
import PropTypes from 'prop-types';
import {base64} from '@ciscospark/common';

import './at-mention.css';

function AtMentionComponent({
  content,
  onEvent
}) {

  function isMention(target) {
    return target.tagName === `SPARK-MENTION` && target.dataset && typeof onEvent === `function`;
  }

  function handleClick(e) {
    triggerEvent(e);
  }

  function handleKeyUp(e) {
    if (e.keyCode && e.keyCode === 13) {
      return triggerEvent(e);
    }
    return false;
  }

  function triggerEvent(e) {
    const target = e.target;
    if (isMention(target)) {
      onEvent({
        resource: `people`,
        action: `clicked`,
        data: {
          personId: constructHydraId(target.dataset.objectType, target.dataset.objectId)
        }
      });
    }
  }

  return ( // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      dangerouslySetInnerHTML={{__html: content}} // eslint-disable-line react/no-danger
      onClick={handleClick}
      onKeyUp={handleKeyUp}
      role="presentation"
    />
  );
}

AtMentionComponent.propTypes = {
  content: PropTypes.string,
  onEvent: PropTypes.func
};

export default (activity, onEvent) => {
  if (activity.mentions) {
    const component = ( // eslint-disable-line no-extra-parens
      <AtMentionComponent
        content={activity.content}
        mentions={activity.mentions.items}
        onEvent={onEvent}
      />
    );

    return Object.assign({}, activity, {component});
  }
  return activity;
};

function constructHydraId(type, id) {
  return base64.encode(`ciscospark://us/${type.toUpperCase()}/${id}`);
}
