// @ciscospark/widget-meet/src/container.js
import React, {PropTypes, Component} from 'react';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';
import classNames from 'classnames';

import Video from '@ciscospark/react-component-video';
import {
  callUser
} from '@ciscospark/redux-module-media';

import styles from './styles.css';

export class MeetWidget extends Component {
  componentDidMount() {
    const props = this.props;
    const {
      shouldStartMeeting,
      toPersonId,
      spark
    } = this.props;
    const {
      media
    } = props;
    if (shouldStartMeeting
      && !media.getIn([`status`, `isCalling`])
      && !media.getIn([`status`, `isConnected`])) {
      props.callUser({userId: toPersonId}, spark);
    }
  }

  shouldComponentUpdate(nextProps) {
    return nextProps !== this.props;
  }

  render() {
    const props = this.props;
    const {
      media
    } = props;
    const remoteUrl = media.getIn([`remoteStream`, `url`]);
    return (
      <div className={classNames(styles.meetContainer, `meet-container`)}>
        <div className={classNames(styles.videoContainer, `video-container`)}>
          {
            remoteUrl &&
            <Video src={remoteUrl} />
          }
        </div>
      </div>
    );
  }
}

MeetWidget.propTypes = {
  shouldStartMeeting: PropTypes.bool,
  spark: PropTypes.object.isRequired,
  toPersonId: PropTypes.string.isRequired
};

function mapStateToProps(state, ownProps) {
  return {
    spark: ownProps.spark || state.spark,
    media: state.media
  };
}

export default connect(
  mapStateToProps,
  (dispatch) => bindActionCreators({
    callUser
  }, dispatch)
)(MeetWidget);
