// @ciscospark/widget-meet/src/messages.js
/*
 * WidgetMeet Messages
 *
 * This contains all the text for the FeaturePage component.
 */
import {defineMessages} from 'react-intl';

export default defineMessages({
  answerButtonLabel: {
    id: `ciscospark.container.meet.button.answer`,
    defaultMessage: `Answer`
  },
  callButtonLabel: {
    id: `ciscospark.container.meet.button.call`,
    defaultMessage: `Call`
  },
  declineButtonLabel: {
    id: `ciscospark.container.meet.button.decline`,
    defaultMessage: `Decline`
  },
  hangupButtonLabel: {
    id: `ciscospark.container.meet.button.hangup`,
    defaultMessage: `Hangup`
  },
  callPersonPrefix: {
    id: `ciscospark.container.meet.message.callPerson`,
    defaultMessage: `Call`
  },
  incomingCallMessage: {
    id: `ciscospark.container.meet.message.incomingCall`,
    defaultMessage: `Incoming call`
  },
  noWebRTCBrowserSupportMessage: {
    id: `ciscospark.container.meet.message.noWebRTCBrowserSupport`,
    defaultMessage: `Video calling is not yet supported in your browser.`
  },
  useSupportedBrowserMessage: {
    id: `ciscospark.container.meet.message.useSupportedBrowser`,
    defaultMessage: `For the best experience, use the latest version of Chrome or Firefox.`
  }
});
