// @ciscospark/widget-meet/src/messages.js
/*
 * WidgetMeet Messages
 *
 * This contains all the text for the FeaturePage component.
 */
import {defineMessages} from 'react-intl';

export default defineMessages({
  answerButtonLabel: {
    id: `ciscospark.container.meet.button.answer`,
    defaultMessage: `Answer`
  },
  callButtonLabel: {
    id: `ciscospark.container.meet.button.call`,
    defaultMessage: `Call`
  },
  declineButtonLabel: {
    id: `ciscospark.container.meet.button.decline`,
    defaultMessage: `Decline`
  },
  hangupButtonLabel: {
    id: `ciscospark.container.meet.button.hangup`,
    defaultMessage: `Hangup`
  },
  okButtonLabel: {
    id: `ciscospark.container.meet.button.ok`,
    defaultMessage: `Ok`
  },
  callErrorDeviceNotFound: {
    id: `ciscospark.container.meet.message.deviceNotFound`,
    defaultMessage: `No calling devices found`
  },
  callErrorNotAllowed: {
    id: `ciscospark.container.meet.message.deviceNotFound`,
    defaultMessage: `Media access has been blocked by browser`
  },
  callErrorUnknownMessage: {
    id: `ciscospark.container.meet.message.errorUnknown`,
    defaultMessage: `An unknown error occurred, please try again`
  },
  callPersonPrefix: {
    id: `ciscospark.container.meet.message.callPerson`,
    defaultMessage: `Call`
  },
  incomingCallMessage: {
    id: `ciscospark.container.meet.message.incomingCall`,
    defaultMessage: `Incoming call`
  },
  noWebRTCBrowserSupportMessage: {
    id: `ciscospark.container.meet.message.noWebRTCBrowserSupport`,
    defaultMessage: `Video calling is not yet supported in your browser.`
  },
  unableToCallMessage: {
    id: `ciscospark.container.meet.message.unableToCallMessage`,
    defaultMessage: `Unable to connect call`
  },
  useSupportedBrowserMessage: {
    id: `ciscospark.container.meet.message.useSupportedBrowser`,
    defaultMessage: `For the best experience, use the latest version of Chrome or Firefox.`
  }
});
