// @ciscospark/widget-meet/src/components/call-incoming/index.js
import React, {PropTypes} from 'react';
import {intlShape} from 'react-intl';
import classNames from 'classnames';

import Avatar from '@ciscospark/react-component-avatar';
import CallControls from '../call-controls';
import messages from '../../messages';

import styles from './styles.css';

export default function IncomingCall({
  avatarImage,
  intl,
  onAnswerClick,
  onDeclineClick,
  toPersonName
}) {
  const {formatMessage} = intl;
  return (
    <div className={classNames(styles.callInactiveContainer, `call-inactive-container`)}>
      <div className={classNames(styles.callAvatar, `call-avatar`)}>
        <Avatar
          image={avatarImage}
          name={toPersonName}
        />
      </div>
      <div className={classNames(styles.personName, `call-person-name`)}>
        {toPersonName}
      </div>
      <div className={classNames(styles.incomingCallLabel, `incoming-call-label`)}>
        {formatMessage(messages.incomingCallMessage)}
      </div>
      <div className={classNames(styles.callControls, `call-controls-container`)}>
        <CallControls
          intl={intl}
          onAnswerClick={onAnswerClick}
          onDeclineClick={onDeclineClick}
          showAnswerButton
          showDeclineButton
          showLabels
        />
      </div>
    </div>
  );
}

IncomingCall.propTypes = {
  avatarImage: PropTypes.string,
  intl: intlShape.isRequired,
  onAnswerClick: PropTypes.func,
  onDeclineClick: PropTypes.func,
  toPersonName: PropTypes.string.isRequired
};
