// @ciscospark/widget-meet/src/components/call-incoming/index.js
import React, {PropTypes} from 'react';
import {intlShape} from 'react-intl';
import classNames from 'classnames';

import Avatar from '@ciscospark/react-component-avatar';
import ButtonControls from '@ciscospark/react-component-button-controls';
import {ICON_TYPE_VIDEO_OUTLINE, ICON_TYPE_EXIT} from '@ciscospark/react-component-icon';
import messages from '../../messages';

import styles from './styles.css';

export default function IncomingCall({
  avatarImage,
  intl,
  onAnswerClick,
  onDeclineClick,
  toPersonName
}) {
  const {formatMessage} = intl;
  const buttons = [
    {
      label: formatMessage(messages.answerButtonLabel),
      iconType: ICON_TYPE_VIDEO_OUTLINE,
      buttonClassName: styles.answerButton,
      onClick: onAnswerClick
    },
    {
      label: formatMessage(messages.declineButtonLabel),
      iconType: ICON_TYPE_EXIT,
      buttonClassName: styles.declineButton,
      onClick: onDeclineClick
    }
  ];

  return (
    <div className={classNames(styles.callInactiveContainer, `call-inactive-container`)}>
      <div className={classNames(styles.callAvatar, `call-avatar`)}>
        <Avatar
          image={avatarImage}
          name={toPersonName}
        />
      </div>
      <div className={classNames(styles.personName, `call-person-name`)}>
        {toPersonName}
      </div>
      <div className={classNames(styles.incomingCallLabel, `incoming-call-label`)}>
        {formatMessage(messages.incomingCallMessage)}
      </div>
      <div className={classNames(styles.callControls, `call-controls-container`)}>
        <ButtonControls
          buttons={buttons}
          showLabels
        />
      </div>
    </div>
  );
}

IncomingCall.propTypes = {
  avatarImage: PropTypes.string,
  intl: intlShape.isRequired,
  onAnswerClick: PropTypes.func,
  onDeclineClick: PropTypes.func,
  toPersonName: PropTypes.string.isRequired
};
