// @ciscospark/react-component-utils/src/constants.js
export const MENTION_NOTIFICATIONS_ON = 'MENTION_NOTIFICATIONS_ON';
export const MENTION_NOTIFICATIONS_OFF = 'MENTION_NOTIFICATIONS_OFF';
export const MESSAGE_NOTIFICATIONS_ON = 'MESSAGE_NOTIFICATIONS_ON';
export const MESSAGE_NOTIFICATIONS_OFF = 'MESSAGE_NOTIFICATIONS_OFF';

export const NOTIFICATIONS_GLOBAL = 'NOTIFICATIONS_GLOBAL';
export const NOTIFICATIONS_OFF = 'NOTIFICATIONS_OFF';
export const NOTIFICATIONS_MENTIONS = 'NOTIFICATIONS_MENTIONS';
export const NOTIFICATIONS_ALL = 'NOTIFICATIONS_ALL';
export const EMAIL_NOTIFICATIONS_MENTIONS = 'EMAIL_NOTIFICATIONS_MENTIONS';
export const EMAIL_NOTIFICATIONS_ONE_TO_ONES = 'EMAIL_NOTIFICATIONS_ONE_TO_ONES';
export const EMAIL_NOTIFICATIONS_DIGEST_EMAILS = 'EMAIL_NOTIFICATIONS_DIGEST_EMAILS';

export const NOTIFICATIONS_BADGE_NONE = 'NOTIFICATIONS_BADGE_NONE';
export const NOTIFICATIONS_BADGE_UNREAD = 'NOTIFICATIONS_BADGE_UNREAD';
export const NOTIFICATIONS_BADGE_MENTION = 'NOTIFICATIONS_BADGE_MENTION';
export const NOTIFICATIONS_BADGE_MUTE = 'NOTIFICATIONS_BADGE_MUTE';
export const NOTIFICATIONS_BADGE_UNMUTE = 'NOTIFICATIONS_BADGE_UNMUTE';

export const FEATURES_USER = 'user';
export const FEATURES_GROUP_MESSAGE_NOTIFICATIONS = 'group-message-notifications';
export const FEATURES_MENTION_NOTIFICATIONS = 'mention-notifications';

export const SPACE_TYPE_ONE_ON_ONE = 'direct';
export const SPACE_TYPE_GROUP = 'group';

export const TAG_MUTED = 'MUTED';
