// @ciscospark/widget-meet/src/components/call-inactive/index.js
import React, {PropTypes} from 'react';
import {intlShape} from 'react-intl';
import classNames from 'classnames';

import Avatar from '@ciscospark/react-component-avatar';
import {ICON_TYPE_VIDEO_OUTLINE} from '@ciscospark/react-component-icon';
import ButtonControls from '@ciscospark/react-component-button-controls';

import messages from '../../messages';
import styles from './styles.css';

export default function InactiveCall({
  toPersonAvatar,
  intl,
  onCallClick,
  toPersonName
}) {
  const {formatMessage} = intl;
  const buttons = [
    {
      label: formatMessage(messages.callButtonLabel),
      iconType: ICON_TYPE_VIDEO_OUTLINE,
      buttonClassName: styles.callButton,
      onClick: onCallClick
    }
  ];
  return (
    <div className={classNames(styles.callInactiveContainer, `call-inactive-container`)}>
      <div className={classNames(styles.callAvatar, `call-avatar`)}>
        <Avatar
          image={toPersonAvatar}
          name={toPersonName}
        />
      </div>
      <div className={classNames(styles.personName, `call-person-name`)}>
        {toPersonName}
      </div>
      <div className={classNames(styles.callControls, `call-controls-container`)}>
        <ButtonControls buttons={buttons} showLabels />
      </div>
    </div>
  );
}

InactiveCall.propTypes = {
  toPersonAvatar: PropTypes.object,
  intl: intlShape.isRequired,
  onCallClick: PropTypes.func,
  toPersonName: PropTypes.string.isRequired
};
