// @ciscospark/widget-meet/src/components/call-inactive/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import PresenceAvatar from '@ciscospark/react-container-presence-avatar';
import ButtonControls from '@ciscospark/react-component-button-controls';

import styles from './styles.css';

const propTypes = {
  avatarId: PropTypes.string,
  avatarImage: PropTypes.string,
  callButtonLabel: PropTypes.string,
  displayName: PropTypes.string,
  onCallClick: PropTypes.func
};

const defaultProps = {
  avatarId: '',
  avatarImage: '',
  callButtonLabel: '',
  displayName: '',
  onCallClick: () => {}
};


function InactiveCall({
  avatarId,
  avatarImage,
  onCallClick,
  callButtonLabel,
  displayName
}) {
  const buttons = [
    {
      label: callButtonLabel,
      buttonClassName: styles.callButton,
      buttonType: 'camera',
      onClick: onCallClick
    }
  ];

  return (
    <div className={classNames(styles.callInactiveContainer, 'call-inactive-container')}>
      <PresenceAvatar
        avatarId={avatarId}
        image={avatarImage}
        name={displayName}
        size={84}
      />
      <div className={classNames(styles.personName, 'call-person-name')}>
        {displayName}
      </div>
      <div className={classNames(styles.callControls, 'call-controls-container')}>
        <ButtonControls buttons={buttons} showLabels />
      </div>
    </div>
  );
}

InactiveCall.propTypes = propTypes;
InactiveCall.defaultProps = defaultProps;

export default InactiveCall;
