// @ciscospark/widget-recents/src/messages.js
/*
 * WidgetRecents Messages
 *
 * This contains all the text for the FeaturePage component.
 */
import {defineMessages} from 'react-intl';

export default defineMessages({
  loadingMessage: {
    id: 'ciscospark.container.recents.loading',
    defaultMessage: 'Loading Webex Teams Spaces'
  },
  sharedFile: {
    id: 'ciscospark.container.recents.sharedFile',
    defaultMessage: 'Shared a file'
  },
  unavailable: {
    id: 'ciscospark.container.recents.userUnavailable',
    defaultMessage: '{actorName} was unavailable.'
  },
  addedToSpace: {
    id: 'ciscospark.container.recents.addedToSpace',
    defaultMessage: '{actorName} added {targetName} to this space.'
  },
  answerButtonLabel: {
    id: 'ciscospark.container.recents.button.answer',
    defaultMessage: 'Answer'
  },
  declineButtonLabel: {
    id: 'ciscospark.container.recents.button.decline',
    defaultMessage: 'Decline'
  },
  incomingCallMessage: {
    id: 'ciscospark.container.recents.message.incomingCall',
    defaultMessage: 'Incoming call'
  },
  errorConnection: {
    id: 'ciscospark.container.recents.error.connection',
    defaultMessage: 'Connection unavailable.'
  },
  errorBadToken: {
    id: 'ciscospark.container.recents.error.badtoken',
    defaultMessage: 'Error: Bad or Invalid Access Token'
  },
  unableToLoad: {
    id: 'ciscospark.container.recents.error.unabletoload',
    defaultMessage: 'Unable to Load Recents'
  },
  unknownError: {
    id: 'ciscospark.container.recents.error.unknown',
    defaultMessage: 'There was a problem loading recents'
  },
  reconnecting: {
    id: 'ciscospark.container.recents.error.reconnecting',
    defaultMessage: 'Reconnecting...'
  },
  viewOlderSpacesButtonLabel: {
    id: 'ciscospark.container.recents.button.viewOlderSpaces',
    defaultMessage: 'View Older Spaces'
  }
});
