// @ciscospark/widget-recents/src/messages.js
/*
 * WidgetMessageMeet Messages
 *
 * This contains all the text for the FeaturePage component.
 */
import {defineMessages} from 'react-intl';

export default defineMessages({
  loadingMessage: {
    id: `ciscospark.container.recents.loading`,
    defaultMessage: `Loading Cisco Spark Spaces`
  },
  sharedFile: {
    id: `ciscospark.container.recents.sharedFile`,
    defaultMessage: `Shared a file`
  },
  unavailable: {
    id: `ciscospark.container.recents.userUnavailable`,
    defaultMessage: `was unavailable`
  },
  addedToSpace: {
    id: `ciscospark.container.recents.addedToSpace`,
    defaultMessage: `{actorName} added {targetName} to this space.`
  },
  answerButtonLabel: {
    id: `ciscospark.container.recents.button.answer`,
    defaultMessage: `Answer`
  },
  declineButtonLabel: {
    id: `ciscospark.container.recents.button.decline`,
    defaultMessage: `Decline`
  },
  incomingCallMessage: {
    id: `ciscospark.container.recents.message.incomingCall`,
    defaultMessage: `Incoming call`
  }
});
