// @ciscospark/redux-module-user/src/index.js
export const initialState = {
  avatars: {},
  avatarsInFlight: [],
  currentUser: {},
  isFetchingCurrentUser: false
};

export const ADD_AVATAR = `ADD_AVATAR`;
export const ADD_AVATAR_BEGIN = `ADD_AVATAR_BEGIN`;
export const GET_CURRENT_USER = `GET_CURRENT_USER`;
export const GET_CURRENT_USER_BEGIN = `GET_CURRENT_USER_BEGIN`;

export default function reducer(state = initialState, action) {
  switch (action.type) {
  case ADD_AVATAR:
    {
      const {userId, avatar} = action.payload;
      const avatarObj = {};
      avatarObj[userId] = avatar;
      const avatars = Object.assign({}, state.avatars, avatarObj);
      const avatarsInFlight = state.avatarsInFlight.filter((inFlightId) => inFlightId !== userId);
      return Object.assign({}, state, {
        avatars,
        avatarsInFlight
      });
    }
  case ADD_AVATAR_BEGIN:
    return Object.assign({}, state, {
      avatarsInFlight: [...state.avatarsInFlight, action.payload.userId]
    });
  case GET_CURRENT_USER:
    return Object.assign({}, state, {
      isFetchingCurrentUser: action.payload.isFetching,
      currentUser: action.payload.user
    });
  case GET_CURRENT_USER_BEGIN:
    return Object.assign({}, state, {
      isFetchingCurrentUser: action.payload.isFetching
    });
  default:
    return state;
  }
}

export function getCurrentUser(user) {
  return {
    type: GET_CURRENT_USER,
    payload: {
      isFetching: false,
      user
    }
  };
}

export function getCurrentUserBegin() {
  return {
    type: GET_CURRENT_USER_BEGIN,
    payload: {
      isFetching: true
    }
  };
}

export function fetchCurrentUser(spark) {
  return (dispatch) => {
    dispatch(getCurrentUserBegin());
    return spark.user.get()
      .then((user) => {
        dispatch(getCurrentUser(user));
      });
  };
}

export function fetchAvatarForUserId(userId, spark) {
  return (dispatch) => {
    dispatch(addAvatarBegin(userId));
    return spark.avatar.retrieveAvatarUrl(userId)
      .then((avatarUrl) =>
        dispatch(addAvatar(userId, avatarUrl))
      );
  };
}

function addAvatar(userId, avatar) {
  return {
    type: ADD_AVATAR,
    payload: {
      userId,
      avatar
    }
  };
}

function addAvatarBegin(userId) {
  return {
    type: ADD_AVATAR_BEGIN,
    payload: {
      userId
    }
  };
}
