// @ciscospark/react-container-file-downloader/src/index.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {saveAs} from 'file-saver';
import {autobind} from 'core-decorators';

import {bufferToBlob, getDisplayName} from '@ciscospark/react-component-utils';
import {retrieveSharedFile} from '@ciscospark/redux-module-share';


export default function injectFileDownloader(WrappedComponent) {
  class FileDownloader extends Component {
    componentDidMount() {
      const {files} = this.props;
      files.forEach((file) => {
        this.getThumbnailImage(file);
      });
    }

    @autobind
    retrieveFile(file) {
      const props = this.props;
      props.retrieveSharedFile(file, props.spark);
    }

    @autobind
    getSharedFileFromStore(fileUrl) {
      const props = this.props;
      return props.share.get(['files', fileUrl]);
    }

    getThumbnailImage(fileObject) {
      const image = fileObject.image;
      if (fileObject.mimeType === 'image/gif' && fileObject.url) {
        this.retrieveFile(fileObject);
      }
      else if (image && (image.url && !this.getSharedFileFromStore(image.url) || image.scr)) {
        this.retrieveFile(image);
      }
    }

    @autobind
    handleDownloadClick(fileObject) {
      const props = this.props;
      const cachedFile = this.getSharedFileFromStore(fileObject.url);
      if (cachedFile && cachedFile.blob) {
        saveAs(cachedFile.blob, fileObject.displayName);
      }
      else {
        props.retrieveSharedFile(fileObject, props.spark)
          .then((file) => {
            const {blob} = bufferToBlob(file);
            saveAs(blob, file.name);
          });
      }
    }

    render() {
      return <WrappedComponent onDownloadClick={this.handleDownloadClick} {...this.props} />;
    }
  }

  FileDownloader.propTypes = {
    files: PropTypes.array
  };

  FileDownloader.displayName = `FileDownloader(${getDisplayName(WrappedComponent)})`;
  FileDownloader.WrappedComponent = WrappedComponent;

  return connect(
    (state) => ({
      share: state.share,
      spark: state.spark.get('spark')
    }),
    (dispatch) => bindActionCreators({
      retrieveSharedFile
    }, dispatch)
  )(FileDownloader);
}
