// @ciscospark/redux-module-flags/src/actions.js
export const ADD_FLAG = `flags/ADD_FLAG`;
export const STORE_FLAGS = `flags/STORE_FLAGS`;
export const REMOVE_FLAG = `flags/REMOVE_FLAG`;
export const UPDATE_FLAG_STATUS = `flags/UPDATE_FLAG_STATUS`;

const IN_FLIGHT_FLAG_ID = `IN_FLIGHT_FLAG_ID`;
const IN_FLIGHT_FLAG_URL = `IN_FLIGHT_FLAG_URL`;


function updateFlagStatus(status) {
  return {
    type: UPDATE_FLAG_STATUS,
    payload: {
      status
    }
  };
}


function addFlag(flag, error) {
  return {
    type: ADD_FLAG,
    payload: {
      flag,
      error
    }
  };
}

function storeFlags(flags, error) {
  return {
    type: STORE_FLAGS,
    payload: {
      flags,
      error
    }
  };
}

function removeFlag(flag, error) {
  return {
    type: REMOVE_FLAG,
    payload: {
      error,
      flag
    }
  };
}

/**
 * Fetches all of the current user's flags
 *
 * @param {any} sparkInstance
 * @returns {function}
 */
export function fetchFlags(sparkInstance) {
  return (dispatch) => {
    dispatch(updateFlagStatus({isFetching: true}));
    return sparkInstance.internal.flag.list()
      .then((flags) =>
        dispatch(storeFlags(flags))
      )
      .catch((error) =>
        dispatch(storeFlags(null, error))
      );
  };
}

/**
 * Flags a given activity. Updates state immediately then
 * adds flag details given from api
 *
 * @param {any} activity
 * @param {any} spark
 * @returns {function}
 */
export function flagActivity(activity, spark) {
  return (dispatch) => {
    dispatch(addFlag({
      id: IN_FLIGHT_FLAG_ID,
      url: IN_FLIGHT_FLAG_URL,
      activityUrl: activity.url
    }));
    return spark.internal.flag.create(activity)
      .then((flag) =>
        dispatch(addFlag(flag))
      )
      .catch((error) =>
        dispatch(addFlag(null, error))
      );
  };
}

/**
 * Removes a flag from the server. Updates the state immediately
 * but re-adds it if the delete fails
 *
 * @param {any} flag
 * @param {any} spark
 * @returns {function}
 */
export function removeFlagFromServer(flag, spark) {
  return (dispatch) => {
    dispatch(removeFlag(flag));
    return spark.internal.flag.delete(flag)
      .catch((error) => {
        dispatch(removeFlag(flag, error));
      });
  };
}

function mapFlag(flag) {
  return {
    id: flag.id,
    url: flag.url,
    activityUrl: flag[`flag-item`]
  };
}
