// @ciscospark/react-redux-spark-metrics/src/component.js
import {Component, PropTypes} from 'react';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';

import {metrics} from '@ciscospark/react-component-utils';

import {
  updateSparkStatus,
  registerDevice,
  storeSparkInstance
} from './actions';

import createSpark from './spark';

class SparkComponent extends Component {

  componentDidMount() {
    const {
      accessToken
    } = this.props;

    const props = this.props;
    let spark = props.spark.get(`spark`);

    if (!spark) {
      spark = createSpark(accessToken);
      metrics.sendLoadStart(spark);
      metrics.sendMetricsInQueue(spark);
      metrics.sendElapsedTimeMetric(`sparkObject:loaded`, spark);
      props.storeSparkInstance(spark);
    }

    spark.listenToAndRun(spark, `change:canAuthorize`, () => {
      metrics.sendElapsedTimeMetric(`sparkObject:authenticated`, spark);
      props.updateSparkStatus({authenticated: spark.canAuthorize});
    });

    spark.listenToAndRun(spark, `change:isAuthenticating`, () => {
      props.updateSparkStatus({authenticating: spark.isAuthenticating});
    });

    spark.listenToAndRun(spark.device, `change:registered`, () => {
      props.updateSparkStatus({registered: spark.device.registered});
    });

    this.setupDevice(spark, props);
  }

  componentWillReceiveProps(nextProps) {
    const spark = nextProps.spark.get(`spark`);
    this.setupDevice(spark, nextProps);
  }

  shouldComponentUpdate(nextProps) {
    const props = this.props;
    return nextProps.spark !== props.spark;
  }

  setupDevice(spark, props) {
    const {
      authenticated,
      registered,
      registering
    } = props.spark.get(`status`).toJS();

    if (authenticated && !registered && !registering) {
      props.registerDevice(spark);
    }
  }

  render() {
    return null;
  }
}

const injectedPropTypes = {
  registerDevice: PropTypes.func,
  spark: PropTypes.object
};

SparkComponent.propTypes = {
  accessToken: PropTypes.string.isRequired,
  ...injectedPropTypes
};

export default connect(
  (state) => ({
    mercury: state.mercury,
    spark: state.spark
  }),
  (dispatch) => bindActionCreators({
    updateSparkStatus,
    registerDevice,
    storeSparkInstance
  }, dispatch)
)(SparkComponent);
