// @ciscospark/widget-meet/src/enhancers/withDestinationType.js
import {compose, lifecycle} from 'recompose';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';
import {
  deconstructHydraId,
  hydraTypes,
  validateSipUri,
  validateEmail
} from '@ciscospark/react-component-utils';

import {storeMeetDetails} from '../actions';


export const destinationTypes = {
  SIP: 'sipUri',
  EMAIL: 'email',
  USERID: 'userId',
  SPACEID: 'spaceId',
  PSTN: 'pstn' // currently ignoring PSTN
};

const destinationTypeNames = Object.values(destinationTypes);

function storeDestinationType(props) {
  const {
    to,
    call,
    type
  } = props;
  const details = {};

  if (to && !call) {
    let destinationType = type;
    let toValue = to;

    if (!destinationType || !destinationTypeNames.includes(destinationType)) {
      const hydraObject = deconstructHydraId(to);
      if (validateSipUri(to)) {
        destinationType = destinationTypes.SIP;
        toValue = to.replace('sip:', '');
      }
      else if (hydraObject.id && hydraObject.type) {
        if (hydraObject.type === hydraTypes.PEOPLE) {
          destinationType = destinationTypes.USERID;
          toValue = hydraObject.id;
          details.userId = toValue;
        }
        else if (hydraObject.type === hydraTypes.ROOM) {
          destinationType = destinationTypes.SPACEID;
          toValue = hydraObject.id;
          details.spaceId = toValue;
        }
      }
      else if (validateEmail(to)) {
        destinationType = destinationTypes.EMAIL;
      }

      details.toType = destinationType;
      details.toValue = toValue;

      props.storeMeetDetails(details);
    }
  }
}

export default compose(
  connect(
    null,
    (dispatch) => bindActionCreators({
      storeMeetDetails
    }, dispatch)
  ),
  lifecycle({
    componentWillMount() {
      storeDestinationType(this.props);
    }
  })
);
