// @ciscospark/react-component-activity-share-files/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import connectFileDownloader from '@ciscospark/react-container-file-downloader';
import ActivityItemBase from '@ciscospark/react-component-activity-item-base';
import ActivityText from '@ciscospark/react-component-activity-text';
import ShareFile from '@ciscospark/react-component-activity-share-file';
import ShareThumbnail from '@ciscospark/react-component-activity-share-thumbnail';

import styles from './styles.css';

export function ActivityShareFiles(props) {
  const {
    content,
    files,
    isPending,
    onDownloadClick,
    share
  } = props;

  const items = files.map((file) => {
    if (file.image) {
      let objectUrl;
      let isFetching = true;
      if (isPending) {
        objectUrl = file.thumbnail;
        isFetching = false;
      }
      else {
        const thumbnail = file.mimeType === `image/gif` ? share.getIn([`files`, file.url]) : share.getIn([`files`, file.image.url]);
        if (thumbnail) {
          isFetching = thumbnail.get(`isFetching`);
          objectUrl = thumbnail.get(`objectUrl`);
        }
      }

      return (
        <ShareThumbnail
          file={file}
          isFetching={isFetching}
          isPending={isPending}
          key={file.url}
          objectUrl={objectUrl}
          onDownloadClick={onDownloadClick}
        />
      );
    }
    return (
      <ShareFile
        file={file}
        isPending={isPending}
        key={file.url}
        onDownloadClick={onDownloadClick}
      />
    );
  });

  let textItem;
  if (content) {
    textItem = <ActivityText content={content} />;
  }

  return (
    <ActivityItemBase {...props}>
      <div className={classNames(`activity-share-list`, styles.shareList)}>
        {items}
        {textItem}
      </div>
    </ActivityItemBase>
  );
}

ActivityShareFiles.propTypes = {
  content: PropTypes.string,
  files: PropTypes.array,
  isPending: PropTypes.bool,
  onDownloadClick: PropTypes.func,
  share: PropTypes.object
};

export default connectFileDownloader(ActivityShareFiles);
