// @ciscospark/redux-module-media/src/reducer.js
import {fromJS} from 'immutable';

import {
  UPDATE_STATUS,
  UPDATE_CALL_STATE,
  CONNECT_CALL,
  REMOVE_CALL,
  STORE_CALL,
  CHECK_WEBRTC_SUPPORT,
  UPDATE_WEBRTC_SUPPORT
} from './actions';

const initialCallObject = fromJS({
  instance: null,
  spaceId: null,
  callState: {},
  startTime: null
});

export const initialState = fromJS({
  calls: {},
  webRTC: {
    hasCheckedSupport: false,
    isSupported: null
  },
  status: {
    isListening: false
  }
});


// eslint-disable-next-line complexity
export default function reducer(state = initialState, action) {
  const {spaceId, call, callState} = action.payload;

  switch (action.type) {

  case UPDATE_STATUS:
    return state.mergeIn([`status`], action.payload.status);

  case STORE_CALL: {
    const callObject = initialCallObject
      .set(`spaceId`, spaceId)
      .set(`instance`, call)
      .mergeIn([`callState`], callState);
    return state.setIn([`calls`, call.locus.id], callObject);
  }

  case UPDATE_CALL_STATE:
    return state.mergeIn([`calls`, call.locus.id, `callState`], callState);

  case CONNECT_CALL:
    return state
      .mergeIn([`calls`, call.locus.id, `callState`], callState)
      .setIn([`calls`, call.locus.id, `callStartTime`], Date.parse(call.locus.fullState.lastActive));

  case REMOVE_CALL:
    return state.deleteIn([`calls`, call.locus.id]);

  case CHECK_WEBRTC_SUPPORT:
    return state.setIn([`webRTC`, `isCheckingSupport`], true);

  case UPDATE_WEBRTC_SUPPORT:
    return state.setIn([`webRTC`, `isSupported`], action.payload.supported);

  default:
    return state;
  }
}
