// @ciscospark/react-component-people-list/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Person from './list-item';
import styles from './styles.css';

const propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    people: PropTypes.arrayOf(PropTypes.shape({
      displayName: PropTypes.string,
      emailAddress: PropTypes.string,
      id: PropTypes.string,
      isExternal: PropTypes.bool,
      isPending: PropTypes.bool
    }))
  })).isRequired,
  onItemClick: PropTypes.func
};

const defaultProps = {
  onItemClick: () => {}
};


function PeopleList({
  items,
  onItemClick
}) {
  if (!items || !items.length) {
    return null;
  }

  function generateGroupList(people) {
    return people.map((p) => {
      const handleItemClick = () => {
        onItemClick(p);
      };

      return (
        <Person
          displayName={p.displayName || p.name}
          key={p.id}
          onClick={handleItemClick}
          {...p}
        />
      );
    });
  }

  return (
    <div>
      {
        items.map((group) => (
          <div className={classNames('ciscospark-people-group', styles.group)} key={group.label || '1'}>
            {
              group.label
              && <h3 className={classNames('ciscospark-people-group-title', styles.title)}>{group.label}</h3>
            }
            {group.people && group.people.length && generateGroupList(group.people)}
          </div>
        ))
      }
    </div>
  );
}

PeopleList.propTypes = propTypes;
PeopleList.defaultProps = defaultProps;

export default PeopleList;
