// @ciscospark/react-component-people-list/src/list-item.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import PresenceAvatar from '@ciscospark/react-container-presence-avatar';

import styles from './list-item.css';


export default function Person({
  displayName,
  id,
  isExternal,
  isPending,
  onClick
}) {

  function handleClick() {
    handleAction();
  }

  function handleKeyPress(e) {
    if (e.key === `Enter`) {
      handleAction();
    }
  }

  function handleAction() {
    if (!isPending) {
      onClick();
    }
  }

  return (
    <div // eslint-disable-line jsx-a11y/no-static-element-interactions
      className={classNames(
        `ciscospark-people-list-item`,
        styles.item,
        {
          [styles.external]: isExternal,
          'ciscospark-people-list-item-external': isExternal,
          [styles.pending]: isPending,
          'ciscospark-people-list-item-pending': isPending
        }
      )}
      onClick={handleClick}
      onKeyPress={handleKeyPress}
      role="button"
      tabIndex="0"
      title={displayName}
    >
      <div className={classNames(`ciscospark-people-list-avatar`, styles.avatar)}>
        <PresenceAvatar avatarId={id} name={displayName} />
      </div>
      <div className={classNames(`ciscospark-people-list-name`, styles.name)}>
        {displayName}
      </div>
    </div>
  );
}

Person.propTypes = {
  displayName: PropTypes.string,
  id: PropTypes.string,
  isExternal: PropTypes.bool,
  isPending: PropTypes.bool,
  onClick: PropTypes.func
};
