// @ciscospark/react-component-people-list/src/list-item.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import PresenceAvatar from '@ciscospark/react-container-presence-avatar';

import styles from './list-item.css';

const propTypes = {
  displayName: PropTypes.string,
  id: PropTypes.string,
  isExternal: PropTypes.bool,
  isPending: PropTypes.bool,
  onClick: PropTypes.func
};

const defaultProps = {
  id: '',
  displayName: '',
  isExternal: false,
  isPending: false,
  onClick: () => {}
};

function Person({
  displayName,
  id,
  isExternal,
  isPending,
  onClick
}) {
  function handleAction() {
    if (!isPending) {
      onClick();
    }
  }

  function handleClick() {
    handleAction();
  }

  function handleKeyPress(e) {
    if (e.key === 'Enter') {
      handleAction();
    }
  }


  return (
    <div
      className={classNames(
        'ciscospark-people-list-item',
        styles.item,
        {
          [styles.external]: isExternal,
          'ciscospark-people-list-item-external': isExternal,
          [styles.pending]: isPending,
          'ciscospark-people-list-item-pending': isPending
        }
      )}
      onClick={handleClick}
      onKeyPress={handleKeyPress}
      role="button"
      tabIndex="0"
      title={displayName}
    >
      <div className={classNames('ciscospark-people-list-avatar', styles.avatar)}>
        <PresenceAvatar avatarId={id} name={displayName} />
      </div>
      <div className={classNames('ciscospark-people-list-name', styles.name)}>
        {displayName}
      </div>
    </div>
  );
}

Person.propTypes = propTypes;
Person.defaultProps = defaultProps;

export default Person;
