// @ciscospark/react-component-call-controls/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';
import {ICON_TYPE_VIDEO_OUTLINE, ICON_TYPE_EXIT} from '@ciscospark/react-component-icon';
import Button from '@ciscospark/react-component-button';

import styles from './styles.css';

export default function CallControls(props) {
  const {
    onAnswerClick,
    onCallClick,
    onHangupClick,
    showAnswerButton,
    showCallButton,
    showHangupButton
  } = props;

  return (
    <div className={`controls-container`}>
      {
        showCallButton &&
        <Button
          buttonClassName={classNames(styles.button, styles.callButton)}
          iconType={ICON_TYPE_VIDEO_OUTLINE}
          onClick={onCallClick}
        />
      }
      {
        showAnswerButton &&
        <Button
          buttonClassName={classNames(styles.button, styles.answerButton)}
          iconType={ICON_TYPE_VIDEO_OUTLINE}
          onClick={onAnswerClick}
        />
      }
      {
        showHangupButton &&
        <Button
          buttonClassName={classNames(styles.button, styles.hangupButton)}
          iconType={ICON_TYPE_EXIT}
          onClick={onHangupClick}
        />
      }
    </div>
  );
}

CallControls.propTypes = {
  onAnswerClick: PropTypes.func,
  onCallClick: PropTypes.func,
  onHangupClick: PropTypes.func,
  showAnswerButton: PropTypes.bool,
  showCallButton: PropTypes.bool,
  showHangupButton: PropTypes.bool
};
