// @ciscospark/react-component-activity-post/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import ActivityItemBase from '@ciscospark/react-component-activity-item-base';
import ActivityText from '@ciscospark/react-component-activity-text';
import AdaptiveCard from '@ciscospark/react-component-adaptive-card';

const propTypes = {
  content: PropTypes.string,
  displayName: PropTypes.string,
  renderedComponent: PropTypes.element,
  cards: PropTypes.array,
  sparkInstance: PropTypes.object.isRequired,
  renderAdaptiveCard: PropTypes.bool
};

const defaultProps = {
  content: '',
  displayName: '',
  renderedComponent: null,
  cards: [],
  renderAdaptiveCard: false
};

function ActivityPost(props) {
  const {
    content,
    renderedComponent,
    displayName,
    renderAdaptiveCard,
    cards,
    sparkInstance
  } = props;

  return (
    <ActivityItemBase {...props}>
      {renderAdaptiveCard
            ? <AdaptiveCard cards={cards} displayName={displayName} sparkInstance={sparkInstance} />
      : <ActivityText content={content} displayName={displayName} renderedComponent={renderedComponent} />}
    </ActivityItemBase>
  );
}

ActivityPost.propTypes = propTypes;
ActivityPost.defaultProps = defaultProps;

export default ActivityPost;
