// @ciscospark/spark-widget-base/src/enhancers/withDOMEvents.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import Events from 'ampersand-events';
import {toClass} from 'recompose';


function formatEventDetails({name, data}) {
  const [resource, eventName] = name.split(':');
  const details = {
    resource,
    event: eventName,
    data
  };

  if (data.actorId) {
    details.actorId = data.actorId;
  }

  if (data.action) {
    details.action = data.action;
  }

  return details;
}


const propTypes = {
  onEvent: PropTypes.func
};

const defaultProps = {
  onEvent: undefined
};


export default function withDOMEvents() {
  return (BaseComponent) => {
    const WrappedComponent = toClass(BaseComponent);

    class WithDOMEvents extends Component {
      constructor(props) {
        super(props);
        this.handleEvent = this.handleEvent.bind(this);
        this.getEl = this.getEl.bind(this);
      }

      getEl(el) {
        this.el = el;
      }

      handleEvent(name, data) {
        const {onEvent} = this.props;
        const detail = formatEventDetails({name, data});
        // Dispatch DOM event
        const event = new CustomEvent(name, {
          detail
        });
        this.el.dispatchEvent(event);

        // Trigger ampersand events
        this.trigger(name, detail);

        // Trigger onEvent callback
        if (typeof onEvent === 'function') {
          return onEvent(name, detail);
        }

        return this;
      }

      render() {
        return (
          <WrappedComponent ref={this.getEl} onEvent={this.handleEvent} {...this.props} />
        );
      }
    }

    WithDOMEvents.propTypes = propTypes;
    WithDOMEvents.defaultProps = defaultProps;

    Object.assign(WithDOMEvents.prototype, Events);

    return WithDOMEvents;
  };
}
