// @webex/widget-meetings/src/components/InactiveMeeting.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import PresenceAvatar from '@ciscospark/react-container-presence-avatar';
import ButtonControls from '@ciscospark/react-component-button-controls';

import styles from './InactiveMeeting.css';

const propTypes = {
  avatarId: PropTypes.string,
  avatarImage: PropTypes.string,
  joinButtonAriaLabel: PropTypes.string,
  joinButtonLabel: PropTypes.string,
  displayName: PropTypes.string,
  onJoinClick: PropTypes.func
};

const defaultProps = {
  avatarId: '',
  avatarImage: '',
  joinButtonAriaLabel: 'Join Meeting',
  joinButtonLabel: 'Join',
  displayName: 'Unknown',
  onJoinClick: () => {}
};


function InactiveMeeting({
  avatarId,
  avatarImage,
  onJoinClick,
  joinButtonAriaLabel,
  joinButtonLabel,
  displayName
}) {
  const buttons = [
    {
      label: joinButtonLabel,
      accessibilityLabel: joinButtonAriaLabel,
      buttonClassName: styles.meetingButton,
      buttonType: 'camera',
      onClick: onJoinClick
    }
  ];

  return (
    <div className={classNames(styles.meetingInactiveContainer, 'meeting-inactive-container')}>
      <PresenceAvatar
        avatarId={avatarId}
        image={avatarImage}
        name={displayName}
        size={84}
      />
      <div className={classNames(styles.personName, 'meeting-person-name')}>
        {displayName}
      </div>
      <div className={classNames(styles.meetingControls, 'meeting-controls-container')}>
        <ButtonControls buttons={buttons} showLabels />
      </div>
    </div>
  );
}

InactiveMeeting.propTypes = propTypes;
InactiveMeeting.defaultProps = defaultProps;

export default InactiveMeeting;
