// @ciscospark/widget-roster/src/components/participant-list-item/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Avatar from '@ciscospark/react-component-avatar';

import styles from './styles.css';

export default function ParticipantListItem({
  participant
}) {

  const participantClasses = [
    `ciscospark-participant-list-item`,
    styles.participantListItem
  ];
  if (participant.external) {
    participantClasses.push(`ciscospark-participant-list-item-external`, styles.external);
  }
  return (
    <div className={classNames(participantClasses)}>
      <div className={classNames(`ciscospark-participant-list-avatar`, styles.avatar)}>
        <Avatar image={participant.avatar} name={participant.displayName} />
      </div>
      <div className={classNames(`ciscospark-participant-list-name`, styles.name)}>
        {participant.displayName}
      </div>
    </div>
  );
}

ParticipantListItem.propTypes = {
  participant: PropTypes.shape({
    avatar: PropTypes.string,
    displayName: PropTypes.string,
    external: PropTypes.bool,
    id: PropTypes.string
  }).isRequired
};
