// @ciscospark/redux-module-mercury/src/index.js
import {Map} from 'immutable';

export const SET_CONNECTING = `mercury/SET_CONNECTING`;
export const SET_DISCONNECTING = `mercury/SET_DISCONNECTING`;
export const SET_CONNECTED = `mercury/SET_CONNECTED`;
export const SET_DISCONNECTED = `mercury/SET_DISCONNECTED`;

export const initialState = new Map({
  connected: false,
  connecting: false
});

export default function reducer(state = initialState, action) {
  switch (action.type) {
  case SET_CONNECTING: {
    return state.set(`connecting`, action.payload);
  }
  case SET_CONNECTED: {
    return state.set(`connected`, action.payload);
  }
  default: {
    return state;
  }
  }
}

function updateStatusConnecting(value) {
  return {
    type: SET_CONNECTING,
    payload: value
  };
}

function updateStatusConnected(value) {
  return {
    type: SET_CONNECTED,
    payload: value
  };
}

export function connectToMercury(spark) {
  return (dispatch) => {
    spark.mercury.on(`change:connecting`, () => {
      dispatch(updateStatusConnecting(spark.mercury.connecting));
    });
    spark.mercury.on(`change:connected`, () => {
      dispatch(updateStatusConnected(spark.mercury.connected));
    });
    return spark.mercury.connect();
  };
}
