// @ciscospark/redux-module-mercury/src/index.js
import {Map} from 'immutable';
import {metrics} from '@ciscospark/react-component-utils';

export const SET_CONNECTING = `mercury/SET_CONNECTING`;
export const SET_CONNECTED = `mercury/SET_CONNECTED`;

export const initialState = new Map({
  status: new Map({
    connected: false,
    connecting: false
  })
});

export default function reducer(state = initialState, action) {
  switch (action.type) {
  case SET_CONNECTING: {
    return state.setIn([`status`, `connecting`], action.payload);
  }
  case SET_CONNECTED: {
    return state.setIn([`status`, `connected`], action.payload);
  }
  default: {
    return state;
  }
  }
}

function updateStatusConnecting(value) {
  return {
    type: SET_CONNECTING,
    payload: value
  };
}

function updateStatusConnected(value) {
  return {
    type: SET_CONNECTED,
    payload: value
  };
}

export function connectToMercury(spark) {
  return (dispatch) => {
    metrics.saveElapsedTime(`mercuryConnection:start`);
    metrics.sendTempMetric(`mercuryConnection:start`, spark);
    spark.mercury.on(`change:connecting`, () => {
      dispatch(updateStatusConnecting(spark.mercury.connecting));
    });
    spark.mercury.on(`change:connected`, () => {
      metrics.saveElapsedTime(`mercuryConnection:end`);
      metrics.sendTempMetric(`mercuryConnection:end`, spark);
      metrics.sendDurationMetric(`mercuryConnection:duration`, `mercuryConnection:start`, `mercuryConnection:end`, spark);
      dispatch(updateStatusConnected(spark.mercury.connected));
    });

    return spark.mercury.connect();
  };
}
