// @ciscospark/react-component-activity-text/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';
import {filterSync} from '@ciscospark/helper-html';

import styles from './styles.css';

export default function ActivityText(props) {

  const {
    content,
    displayName
  } = props;

  /* eslint-disable-reason content is considered safe from server */
  /* eslint-disable react/no-danger */
  if (content) {
    const cleaned = filterSync(() => {}, { // eslint-disable-line no-empty-function
      'spark-mention': [`data-object-type`, `data-object-id`, `data-object-url`],
      a: [`href`],
      b: [],
      blockquote: [`class`],
      strong: [],
      i: [],
      em: [],
      pre: [],
      code: [`class`],
      br: [],
      hr: [],
      p: [],
      ul: [],
      ol: [],
      li: [],
      h1: [],
      h2: [],
      h3: [],
      h4: [],
      h5: [],
      h6: []
    }, [], content);
    const htmlContent = {__html: cleaned};
    return (
      <div
        className={classNames(`ciscospark-activity-text`, styles.activityText)}
        dangerouslySetInnerHTML={htmlContent}
      />
    );
  }

  return (
    <div className={classNames(`ciscospark-activity-text`, styles.activityText)} >
      {displayName}
    </div>
  );

  /* eslint-enable react/no-danger */
}

ActivityText.propTypes = {
  content: PropTypes.string,
  displayName: PropTypes.string
};
