// @webex/widget-voice-mail/src/ScrubbingBar.tsx
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/destructuring-assignment */
import { SliderState, SliderStateOptions, useSliderState } from 'react-stately';

import React, { MutableRefObject, useRef, useState } from 'react';
import {
  VisuallyHidden,
  mergeProps,
  useFocusRing,
  useNumberFormatter,
  useSlider,
  useSliderThumb,
} from 'react-aria';
import './ScrubbingBar.styles.scss';
import useWebexClasses from './hooks/useWebexClasses';
import { formatSleekBarCurrTimeDurationForAnnouncement, formatVoiceMailTimeDurationForAnnouncement } from '@webex/widget-call-history/src/utils/voiceOver';
import { useTranslation } from 'react-i18next';

type IScrubbingBarThumbProps = {
  state: SliderState;
  index: number;
  trackRef: MutableRefObject<HTMLDivElement | null>;
  className?: string;
  currTime?: number[];
};

/**
 *
 * @param props
 * @param props.state
 * @param props.trackRef
 * @param props.index
 * @param props.className
 */
const ScrubbingBarThumb = ({
  state,
  trackRef,
  index,
  className = undefined,
  currTime
}: IScrubbingBarThumbProps) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const [isFocused, setIsFocused] = useState(false);
  const { thumbProps, inputProps } = useSliderThumb(
    {
      index,
      trackRef,
      inputRef,
    },
    state
  );

  const { focusProps } = useFocusRing();
  const handleFocus = () => {
    setIsFocused(true);
  };

  const handleBlur = () => {
    setIsFocused(false);
  };
  return (
    <div
      data-testid="scrubbing-bar-thumb"
      {...mergeProps(thumbProps, focusProps)}
      className={`${className} ${isFocused ? 'vm-focus-ring-wrapper' : ''}`}
      style={{
        left: `calc(calc(100% - 16px) * ${state.getThumbPercent(index)})`,
      }}
      tabIndex={0}
      onFocus={handleFocus}
      onBlur={handleBlur}
      role="slider"
      aria-valuenow={currTime ? currTime[0] : 0}
      aria-valuetext={currTime ? `${formatSleekBarCurrTimeDurationForAnnouncement(currTime[0] * 1000)}` : ""}
    >
      <VisuallyHidden>
        <input ref={inputRef} {...inputProps} />
      </VisuallyHidden>
    </div>
  );
};

export interface IScrubbingBarProps
  extends Partial<SliderStateOptions<number[]>> {
  numberFormatOptions?: unknown;
  duration?: number;
  voicemailName?: string;
}

export const ScrubbingBar = ({ ...rest }: IScrubbingBarProps) => {
  const [cssClasses, sc] = useWebexClasses(
    'voicemail-scrubbing-bar',
    undefined,
    {
      disabled: !!rest.isDisabled,
    }
  );
  const trackRef = React.useRef(null);
  const numberFormatted = useNumberFormatter();
  const state = useSliderState({
    numberFormatter: numberFormatted,
    ...rest,
  });
  const { groupProps, trackProps } = useSlider(rest, state, trackRef);
  const { t } = useTranslation('WebexVoicemail');

  return (
    <div {...groupProps} className={cssClasses} data-testid="scrubbing-bar" aria-label={`${t('voicemailFrom')} ${rest.voicemailName}, ${t('slider')}, 0 minute 0 second of ${rest.duration ? (formatVoiceMailTimeDurationForAnnouncement(rest.duration as number)) : '0'}`}>
      <div {...trackProps} ref={trackRef} className={sc('track')}>
        <div
          className={sc('viewed')}
          style={{ width: `${state.getThumbPercent(0) * 100}%` }}
        />
        <ScrubbingBarThumb
          index={0}
          state={state}
          trackRef={trackRef}
          className={sc('thumb')}
          currTime={rest.value}
        />
      </div>
    </div>
  );
};
