// @ciscospark/react-component-activity-system-message/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import {FormattedMessage} from 'react-intl';

import messages from './messages';
import styles from './styles.css';

export const SYSTEM_MESSAGE_VERB_TOMBSTONE = `tombstone`;
export const SYSTEM_MESSAGE_VERB_CREATE = `create`;
export const SYSTEM_MESSAGE_VERBS = [SYSTEM_MESSAGE_VERB_CREATE, SYSTEM_MESSAGE_VERB_TOMBSTONE];

export default function ActivitySystemMessage(props) {
  const {
    isSelf,
    name,
    timestamp,
    verb
  } = props;

  let systemMessage;
  if (verb === SYSTEM_MESSAGE_VERB_CREATE) {
    if (isSelf) {
      systemMessage = <FormattedMessage {...messages.youCreate} />;
    }
    else {
      systemMessage = <FormattedMessage {...messages.someoneCreate} values={{name}} />;
    }
  }
  else if (verb === SYSTEM_MESSAGE_VERB_TOMBSTONE) {
    if (isSelf) {
      systemMessage = <FormattedMessage {...messages.youDelete} />;
    }
    else {
      systemMessage = <FormattedMessage {...messages.someoneDelete} values={{name}} />;
    }
  }

  return (
    <div className={classNames(`system-message`, styles.systemMessage)}>
      {systemMessage} {timestamp}
    </div>
  );
}

ActivitySystemMessage.propTypes = {
  isSelf: PropTypes.bool,
  name: PropTypes.string.isRequired,
  timestamp: PropTypes.string,
  verb: PropTypes.string.isRequired
};
