// @ciscospark/react-component-activity-system-message/src/messages.js
/*
 * System Message
 *
 * This contains all the text for the ActivitySystemMessage component.
 */
import {defineMessages} from 'react-intl';

export default defineMessages({
  youCreate: {
    id: `ciscospark.systemMessage.conversation.youCreate`,
    defaultMessage: `You created this conversation.`
  },
  someoneCreate: {
    id: `ciscospark.systemMessage.conversation.someoneCreate`,
    defaultMessage: `{name} created this conversation.`
  },
  youDelete: {
    id: `ciscospark.systemMessage.message.youDelete`,
    defaultMessage: `You deleted your message.`
  },
  someoneDelete: {
    id: `ciscospark.systemMessage.message.someoneDelete`,
    defaultMessage: `{name} deleted their own message.`
  },
  someoneWasUnavailable: {
    id: `ciscospark.systemMessage.message.someoneWasUnavailable`,
    defaultMessage: `{name} was unavailable.`
  },
  youCancelledACall: {
    id: `ciscospark.systemMessage.message.youCancelledACall`,
    defaultMessage: `You cancelled a call.`
  },
  someoneCancelledACall: {
    id: `ciscospark.systemMessage.message.someoneCancelledACall`,
    defaultMessage: `{name} cancelled a call.`
  },
  youHadACall: {
    id: `ciscospark.systemMessage.message.youHadACall`,
    defaultMessage: `You had a call ({duration})`
  },
  youMissedACall: {
    id: `ciscospark.systemMessage.message.youMissedACall`,
    defaultMessage: `You missed a call.`
  }
});
