// @ciscospark/react-component-activity-system-message/src/messages.js
/*
 * System Message
 *
 * This contains all the text for the ActivitySystemMessage component.
 */
import {defineMessages} from 'react-intl';

export default defineMessages({
  youCreate: {
    id: `ciscospark.systemMessage.conversation.youCreate`,
    defaultMessage: `You created this conversation.`
  },
  someoneCreate: {
    id: `ciscospark.systemMessage.conversation.someoneCreate`,
    defaultMessage: `{name} created this conversation.`
  },
  youDelete: {
    id: `ciscospark.systemMessage.message.youDelete`,
    defaultMessage: `You deleted your message.`
  },
  someoneDelete: {
    id: `ciscospark.systemMessage.message.someoneDelete`,
    defaultMessage: `{name} deleted their own message.`
  },
  someoneWasUnavailable: {
    id: `ciscospark.systemMessage.message.someoneWasUnavailable`,
    defaultMessage: `{name} was unavailable.`
  },
  youCancelledACall: {
    id: `ciscospark.systemMessage.message.youCancelledACall`,
    defaultMessage: `You cancelled a call.`
  },
  someoneCancelledACall: {
    id: `ciscospark.systemMessage.message.someoneCancelledACall`,
    defaultMessage: `{name} cancelled a call.`
  },
  youHadACall: {
    id: `ciscospark.systemMessage.message.youHadACall`,
    defaultMessage: `You had a call ({duration})`
  },
  youMissedACall: {
    id: `ciscospark.systemMessage.message.youMissedACall`,
    defaultMessage: `You missed a call.`
  },
  youHadAMeeting: {
    id: `ciscospark.systemMessage.message.youHadAMeeting`,
    defaultMessage: `You had a meeting ({duration})`
  },
  nobodyJoinedYourMeeting: {
    id: `ciscospark.systemMessage.message.nobodyJoinedYourMeeting`,
    defaultMessage: `You started a meeting but no one was available.`
  },
  nobodyJoinedTheMeeting: {
    id: `ciscospark.systemMessage.message.nobodyJoinedTheMeeting`,
    defaultMessage: `{name} cancelled a meeting.`
  },
  youMissedAMeeting: {
    id: `ciscospark.systemMessage.message.youMissedAMeeting`,
    defaultMessage: `You missed a meeting ({duration})`
  },
  youAdded: {
    id: `ciscospark.systemMessage.message.youAdded`,
    defaultMessage: `You added {name} to this space.`
  },
  someoneAdded: {
    id: `ciscospark.systemMessage.message.youAdded`,
    defaultMessage: `{name} was added to this space.`
  },
  youRemoved: {
    id: `ciscospark.systemMessage.message.youRemoved`,
    defaultMessage: `You removed {name} from this space.`
  },
  someoneRemoved: {
    id: `ciscospark.systemMessage.message.youRemoved`,
    defaultMessage: `{name} was removed from this space.`
  }
});
