// @ciscospark/react-component-activity-item/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import ActivityPost from '@ciscospark/react-component-activity-post';
import ActivityShareFiles from '@ciscospark/react-component-activity-share-files';
import ActivitySystemMessage, {SYSTEM_MESSAGE_VERBS} from '@ciscospark/react-component-activity-system-message';

import styles from './styles.css';

const POST_VERB = `post`;
const SHARE_VERB = `share`;

export default function ActivityItem(props) {
  const {
    activity,
    verb
  } = props;

  let itemComponent = ``;

  let content;
  if (activity.content) {
    // HTML encoded posts
    content = activity.content;
  }
  else {
    content = activity.displayName;
  }

  if (verb === POST_VERB) {
    itemComponent = <ActivityPost content={content} {...props} />;
  }
  else if (verb === SHARE_VERB) {
    itemComponent = <ActivityShareFiles content={content} files={activity.files.items} {...props} />;
  }
  else if (SYSTEM_MESSAGE_VERBS.indexOf(verb) !== -1) {
    itemComponent = <ActivitySystemMessage {...props} />;
  }

  return (
    <div className={classNames(`activity-item-container`, styles.activityItemContainer)}>
      {itemComponent}
    </div>
  );
}
/* eslint-disable react/no-unused-prop-types */
/* eslint-disable-reason passing props via destructuring */
ActivityItem.propTypes = {
  activity: PropTypes.shape({
    content: PropTypes.string,
    displayName: PropTypes.string,
    files: PropTypes.shape({
      items: PropTypes.array
    })
  }),
  avatarUrl: PropTypes.string,
  callData: PropTypes.shape({
    callInfo: PropTypes.shape({
      duration: PropTypes.string,
      name: PropTypes.string
    }),
    status: PropTypes.string
  }),
  id: PropTypes.string.isRequired,
  isAdditional: PropTypes.bool,
  isFlagged: PropTypes.bool,
  isPending: PropTypes.bool,
  isSelf: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onActivityDelete: PropTypes.func,
  onActivityFlag: PropTypes.func,
  timestamp: PropTypes.string,
  verb: PropTypes.string.isRequired
};
/* eslint-enable react/no-unused-prop-types */
