// @ciscospark/react-component-activity-item/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import ActivityPost from '@ciscospark/react-component-activity-post';
import ActivityShareFiles from '@ciscospark/react-component-activity-share-files';
import ActivitySystemMessage, {SYSTEM_MESSAGE_VERBS} from '@ciscospark/react-component-activity-system-message';

import styles from './styles.css';

const POST_VERB = 'post';
const SHARE_VERB = 'share';

export default function ActivityItem(props) {
  const {
    activity,
    verb
  } = props;

  let itemComponent = '';


  if (verb === POST_VERB) {
    itemComponent = ( // eslint-disable-line no-extra-parens
      <ActivityPost
        content={activity.content}
        displayName={activity.displayName}
        renderedComponent={activity.component}
        {...props}
      />
    );
  }
  else if (verb === SHARE_VERB) {
    itemComponent = ( // eslint-disable-line no-extra-parens
      <ActivityShareFiles
        content={activity.content}
        displayName={activity.displayName}
        files={activity.files.items}
        {...props}
      />
    );
  }
  else if (SYSTEM_MESSAGE_VERBS.indexOf(verb) !== -1) {
    itemComponent = <ActivitySystemMessage {...props} />;
  }

  return (
    <div className={classNames('ciscospark-activity-item-container', styles.activityItemContainer)}>
      {itemComponent}
    </div>
  );
}
/* eslint-disable react/no-unused-prop-types */
/* eslint-disable-reason passing props via destructuring */
ActivityItem.propTypes = {
  activity: PropTypes.shape({
    content: PropTypes.string,
    displayName: PropTypes.string,
    files: PropTypes.shape({
      items: PropTypes.array
    })
  }),
  actorId: PropTypes.string,
  callData: PropTypes.shape({
    callInfo: PropTypes.shape({
      duration: PropTypes.string,
      name: PropTypes.string
    }),
    status: PropTypes.string
  }),
  hasError: PropTypes.bool,
  id: PropTypes.string.isRequired,
  isAdditional: PropTypes.bool,
  isFlagged: PropTypes.bool,
  isPending: PropTypes.bool,
  isSelf: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onActivityDelete: PropTypes.func,
  onActivityFlag: PropTypes.func,
  onActivityRetry: PropTypes.func,
  timestamp: PropTypes.string,
  verb: PropTypes.string.isRequired
};
/* eslint-enable react/no-unused-prop-types */
