// @ciscospark/react-component-avatar/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon, {ICON_TYPE_MESSAGE} from '@ciscospark/react-component-icon';

import styles from './styles.css';

const propTypes = {
  baseColor: PropTypes.string,
  image: PropTypes.string,
  isSelfAvatar: PropTypes.bool,
  name: PropTypes.string,
  size: PropTypes.oneOf(['small', 'medium', 'large'])
};

const defaultProps = {
  baseColor: '',
  image: '',
  isSelfAvatar: false,
  name: '',
  size: 'medium'
};

function Avatar({
  baseColor,
  name,
  image,
  isSelfAvatar,
  size
}) {
  let avatarClass, avatarContents;
  let backgroundStyle = {};

  const letter = name ? name.replace(/[^0-9a-z]/gi, '').substr(0, 1).toUpperCase() : undefined;

  if (isSelfAvatar) {
    avatarContents = <span className={classNames('ciscospark-avatar-self', styles.avatarSelf)}><Icon type={ICON_TYPE_MESSAGE} /></span>;
  }
  else if (image) {
    backgroundStyle = {backgroundImage: `url('${image}')`};
  }
  else if (baseColor) {
    backgroundStyle.backgroundColor = baseColor;
    avatarClass = [styles.avatarColorLetter, styles.avatarLetter];
    avatarContents = <span>{letter}</span>;
  }
  else if (name) {
    avatarClass = styles.avatarLetter;
    avatarContents = <span>{letter}</span>;
  }

  return (
    <div className={classNames('ciscospark-avatar', styles.avatar, `avatar-${size}`, styles[size], avatarClass)} style={backgroundStyle}>
      {avatarContents}
    </div>
  );
}

Avatar.propTypes = propTypes;
Avatar.defaultProps = defaultProps;

export default Avatar;
