// @ciscospark/react-component-avatar/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon, {ICON_TYPE_MESSAGE} from '@ciscospark/react-component-icon';
import styles from './styles.css';

function Avatar({baseColor, name, image, isSelfAvatar, size}) {
  let avatarClass, avatarContents;
  let backgroundStyle = {};


  if (baseColor) {
    backgroundStyle.backgroundColor = baseColor;
    avatarClass = [styles.avatarColorLetter, styles.avatarLetter];
    avatarContents = <span>{name.substr(0, 1)}</span>;
  }
  else if (isSelfAvatar) {
    avatarContents = <span className={classNames(`ciscospark-avatar-self`, styles.avatarSelf)}><Icon type={ICON_TYPE_MESSAGE} /></span>;
  }
  else if (image) {
    backgroundStyle = {backgroundImage: `url('${image}')`};
  }
  else if (name) {
    const letter = name.split(` `).slice(0, 2).map((n) => n.substr(0, 1)).join(``);
    avatarClass = styles.avatarLetter;
    avatarContents = <span>{letter}</span>;
  }

  return (
    <div className={classNames(`ciscospark-avatar`, styles.avatar, `avatar-${size}`, styles[size], avatarClass)} style={backgroundStyle}>
      {avatarContents}
    </div>
  );
}

Avatar.propTypes = {
  baseColor: PropTypes.string,
  image: PropTypes.string,
  isSelfAvatar: PropTypes.bool,
  name: PropTypes.string,
  size: PropTypes.oneOf([`small`, `medium`, `large`])
};

export default Avatar;
