// @ciscospark/react-component-button-controls/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {ActivityButton} from '@collab-ui/react';

import styles from './styles.css';

const propTypes = {
  buttons: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    buttonClassName: PropTypes.string,
    buttonType: PropTypes.oneOfType([
      PropTypes.oneOf(['chat', 'camera', 'meetings', 'whiteboard', 'files', 'share-screen', 'tasks']),
      PropTypes.shape({
        color: PropTypes.string,
        icon: PropTypes.element.isRequired
      })
    ]),
    onClick: PropTypes.func
  })).isRequired,
  showLabels: PropTypes.bool
};

const defaultProps = {
  showLabels: false
};

function ButtonControls(props) {
  const {
    buttons,
    showLabels
  } = props;

  const buttonsRendered = buttons.map((button) => {
    const label = showLabels && button.label ? button.label : '';
    return (
      <ActivityButton
        {...button}
        label={label}
        ariaLabel={label}
        type={button.buttonType}
      />
    );
  });

  return (
    <div className={classNames('ciscospark-controls-container', styles.controlContainer)}>
      {buttonsRendered}
    </div>
  );
}

ButtonControls.propTypes = propTypes;
ButtonControls.defaultProps = defaultProps;

export default ButtonControls;
