// @ciscospark/widget-meet/src/components/call-active/index.js
import React from 'react';
import PropTypes from 'prop-types';
import Draggable from 'react-draggable';
import classNames from 'classnames';
import ButtonControls from '@ciscospark/react-component-button-controls';
import Video from '@ciscospark/react-component-video';
import Audio from '@ciscospark/react-component-audio';
import Avatar from '@ciscospark/react-component-avatar';
import {ICONS} from '@ciscospark/react-component-icon';

import withActiveCallHandlers from '../../enhancers/withActiveCallHandlers';

import styles from './styles.css';

const propTypes = {
  avatarImage: PropTypes.string,
  displayName: PropTypes.string,
  localMediaStream: PropTypes.object,
  localVideoPosition: PropTypes.object,
  onHangupClick: PropTypes.func.isRequired,
  onLocalVideoDragStop: PropTypes.func,
  onStartSendingAudio: PropTypes.func.isRequired,
  onStartSendingVideo: PropTypes.func.isRequired,
  onStopSendingAudio: PropTypes.func.isRequired,
  onStopSendingVideo: PropTypes.func.isRequired,
  remoteAudioStream: PropTypes.object,
  remoteMediaStream: PropTypes.object,
  remoteVideoStream: PropTypes.object,
  isSendingAudio: PropTypes.bool,
  isSendingVideo: PropTypes.bool,
  isConnected: PropTypes.bool,
  activeParticipantsCount: PropTypes.number,
  remoteVideoMuted: PropTypes.bool
};

const defaultProps = {
  avatarImage: '',
  displayName: '',
  localMediaStream: {},
  localVideoPosition: {x: 0, y: 0},
  onLocalVideoDragStop: () => {},
  remoteAudioStream: null,
  remoteMediaStream: null,
  remoteVideoStream: null,
  isSendingAudio: false,
  isSendingVideo: false,
  isConnected: false,
  activeParticipantsCount: 0,
  remoteVideoMuted: false
};

function ActiveCall({
  localVideoPosition,
  onHangupClick,
  onLocalVideoDragStop,
  onStartSendingAudio,
  onStartSendingVideo,
  onStopSendingAudio,
  onStopSendingVideo,
  localMediaStream,
  remoteAudioStream,
  remoteMediaStream,
  remoteVideoStream,
  avatarImage,
  displayName,
  remoteVideoMuted,
  isSendingAudio,
  isSendingVideo,
  isConnected,
  activeParticipantsCount
}) {
  const connectedClass = remoteMediaStream && remoteMediaStream.active && styles.callConnected;

  const buttons = [
    {
      accessibilityLabel: 'Mute Audio',
      iconType: ICONS.ICON_TYPE_MUTE_OUTLINE,
      buttonClassName: classNames(styles.audioButton, isSendingAudio === false && isConnected
        ? styles.audioStartSending : styles.audioStopSending),
      onClick: isSendingAudio ? onStopSendingAudio : onStartSendingAudio
    },
    {
      accessibilityLabel: 'Mute Video',
      iconType: ICONS.ICON_TYPE_VIDEO_CROSS_OUTLINE,
      buttonClassName: classNames(styles.videoButton, isSendingVideo === false && isConnected
        ? styles.videoStartSending : styles.videoStopSending),
      onClick: isSendingVideo ? onStopSendingVideo : onStartSendingVideo
    },
    {
      accessibilityLabel: 'Hangup',
      iconType: ICONS.ICON_TYPE_EXIT,
      buttonClassName: styles.hangupButton,
      onClick: onHangupClick
    }
  ];

  let remoteAudio, remoteVideo;
  if (remoteMediaStream && remoteMediaStream.active) {
    remoteAudio = <Audio srcObject={remoteAudioStream} />;
    remoteVideo = <Video srcObject={remoteVideoStream} />;
    if (remoteVideoMuted) {
      remoteVideo = (
        <div className={classNames(styles.avatarView, 'remote-avatar-view')}>
          <Avatar image={avatarImage} name={displayName} />
        </div>
      );
    }
  }

  return (
    <div className={classNames(styles.callContainer, connectedClass, 'call-container')}>
      { activeParticipantsCount > 1 &&
        <div className={classNames(styles.remoteVideo, 'remote-video')}>
          { remoteVideo }
          { remoteAudio }
        </div>
      }
      {
        activeParticipantsCount < 2 &&
        <div className={classNames(styles.waiting, 'participants-waiting')}>Waiting for others to join...</div>
      }
      {
        localMediaStream && localMediaStream.active && (isSendingVideo || isSendingAudio) &&
        <Draggable
          bounds="parent"
          defaultClassNameDragging={styles.reactDraggableDragging}
          onStop={onLocalVideoDragStop}
          position={localVideoPosition}
        >
          <div className={classNames(styles.localVideo, 'local-video')}>
            <Video audioMuted srcObject={localMediaStream} />
          </div>
        </Draggable>
      }
      <div className={classNames(styles.callControls, 'call-controls')}>
        <ButtonControls buttons={buttons} showLabels={false} />
      </div>
    </div>
  );
}

ActiveCall.propTypes = propTypes;
ActiveCall.defaultProps = defaultProps;

export default withActiveCallHandlers(ActiveCall);
