// @ciscospark/widget-recents/src/components/RecentsHeader.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import {Button, Icon, SearchInput} from '@collab-ui/react';

import PresenceAvatar from '@ciscospark/react-container-presence-avatar';


import styles from './RecentsHeader.css';

const propTypes = {
  currentUser: PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string
  }).isRequired,
  enableAddButton: PropTypes.bool.isRequired,
  enableSpaceListFilter: PropTypes.bool.isRequired,
  enableUserProfile: PropTypes.bool.isRequired,
  onAddClick: PropTypes.func,
  onFilterChange: PropTypes.func,
  onProfileClick: PropTypes.func
};

const defaultProps = {
  onAddClick: () => {},
  onProfileClick: () => {},
  onFilterChange: () => {}
};

function RecentsHeader(props) {
  const {
    currentUser,
    enableAddButton,
    enableSpaceListFilter,
    enableUserProfile,
    onAddClick,
    onFilterChange,
    onProfileClick
  } = props;

  return (
    <div className={classNames('ciscospark-recents-header', styles.recentsHeader, 'cui--dark')}>
      <div className={classNames('ciscospark-recents-header-left', styles.headerSideItem)}>
        {
          enableUserProfile &&
          <PresenceAvatar
            avatarId={currentUser.id}
            name={currentUser.displayName}
            onClick={onProfileClick}
          />
        }
      </div>
      <div className={classNames('ciscospark-search-input-wrapper', styles.searchInputWrapper)}>
        {
          enableSpaceListFilter &&
          <div className={classNames('ciscospark-search-input', styles.searchInput)}>
            <SearchInput
              name="pillSearchInput"
              htmlId="pillSearchInput"
              type="pill"
              onKeyUp={onFilterChange}
              onChange={onFilterChange}
              clear
            />
          </div>
        }
      </div>
      <div className={classNames('ciscospark-recents-header-right', styles.headerSideItem)}>
        {
          enableAddButton &&
          <Button
            ariaLabel="Contact a person or create a space"
            onClick={onAddClick}
            circle
            color="white"
            size={40}
            title="Contact a person or create a space"
          >
            <Icon
              name="plus_18"
            />
          </Button>
        }
      </div>
    </div>
  );
}

RecentsHeader.propTypes = propTypes;
RecentsHeader.defaultProps = defaultProps;

export default RecentsHeader;
