// @ciscospark/react-component-file-staging-area/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {bytesToSize, getFileType} from '@ciscospark/react-component-utils';
import Button from '@ciscospark/react-component-button';
import ChipFile from '@ciscospark/react-component-chip-file';

import styles from './styles.css';

const propTypes = {
  files: PropTypes.object.isRequired,
  onFileRemove: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired
};

function FileStagingArea({
  onFileRemove,
  onSubmit,
  files
}) {
  if (!files || !Object.keys(files).length) {
    return null;
  }

  const fileChips = [];

  files.forEach((file) => {
    const chipProps = {
      name: file.name,
      type: getFileType(file.type),
      size: bytesToSize(file.fileSize),
      id: file.id,
      onRemove: onFileRemove,
      thumbnail: file.thumbnail
    };

    const chip = (
      <div className={classNames('ciscospark-chip-container', styles.chipContainer)} key={file.id}>
        <ChipFile {...chipProps} />
      </div>
    );

    fileChips.push(chip);
  });

  return (
    <div className={classNames('ciscospark-file-staging-area', styles.fileStagingArea)}>
      <div className={classNames('ciscospark-staged-files', styles.files)}>
        {fileChips}
      </div>
      <div className={classNames('ciscospark-staging-actions', styles.actions)}>
        <Button label="Share" onClick={onSubmit} />
      </div>
    </div>
  );
}

FileStagingArea.propTypes = propTypes;

export default FileStagingArea;
