// @ciscospark/widget-files/src/container.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {autobind} from 'core-decorators';

import ActivityMenuHeader from '@ciscospark/react-component-activity-menu-header';

import getFilesWidgetProps from './selectors';

const ownPropTypes = {
  onClickClose: PropTypes.func.isRequired,
  onClickMenu: PropTypes.func.isRequired
};

const injectedPropTypes = {
  conversation: PropTypes.object.isRequired,
  fileShares: PropTypes.object.isRequired
};

class FilesWidget extends Component {
  @autobind
  handleMenuClick() {
    this.props.onClickMenu();
  }

  @autobind
  handleCloseClick() {
    this.props.onClickClose();
  }

  render() {
    return (
      <div>
        <ActivityMenuHeader
          onClose={this.handleCloseClick}
          onMenuClick={this.handleMenuClick}
          title={`Files (${this.props.fileShares.length})`}
        />
        {
          this.props.fileShares.length && (
            <div>Files</div>
          )
        }
        {
          !this.props.fileShares.length && (
            <div>No Files Found</div>
          )
        }
      </div>
    );
  }
}

FilesWidget.propTypes = {
  ...ownPropTypes,
  ...injectedPropTypes
};

export default connect(
  getFilesWidgetProps
)(FilesWidget);
