// @ciscospark/widget-recents/src/components/spaces-list/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import SpaceItem from '../space-item';

const propTypes = {
  currentUser: PropTypes.shape({
    id: PropTypes.string.isRequired
  }).isRequired,
  formatMessage: PropTypes.func.isRequired,
  hasGroupCalling: PropTypes.bool,
  onCallClick: PropTypes.func,
  onClick: PropTypes.func,
  spaces: PropTypes.arrayOf(
    PropTypes.shape({
      avatarUrl: PropTypes.string,
      call: PropTypes.object,
      id: PropTypes.string,
      isDecrypting: PropTypes.bool,
      isUnread: PropTypes.bool,
      lastActivityTime: PropTypes.string,
      latestActivity: PropTypes.object,
      name: PropTypes.string,
      teamColor: PropTypes.string,
      teamName: PropTypes.string,
      type: PropTypes.string
    })
  )
};

const defaultProps = {
  hasGroupCalling: false,
  onCallClick: () => {},
  onClick: () => {},
  spaces: []
};

export default function SpacesList({
  currentUser,
  formatMessage,
  hasGroupCalling,
  onCallClick,
  onClick,
  spaces
}) {
  const recents = [];
  spaces.forEach((space) => recents.push(<SpaceItem
    currentUser={currentUser}
    formatMessage={formatMessage}
    hasGroupCalling={hasGroupCalling}
    key={space.id}
    onCallClick={onCallClick}
    onClick={onClick}
    {...space}
  />));

  return (
    <div className={classNames('ciscospark-spaces-list')}>
      {recents}
    </div>
  );
}

SpacesList.propTypes = propTypes;
SpacesList.defaultProps = defaultProps;
