// @webex/widget-speed-dial/src/SpeedDial.utils.tsx
import { FieldError } from 'react-hook-form';
import { Message } from '@momentum-ui/react-collaboration/dist/esm/components/InputMessage/InputMessage.types';
import { IFormData, ISelectItems, PhoneType, PhoneValueTypes } from './SpeedDialForm.types';

/**
 * Handles converting field error to message.
 *
 * @param {FieldError} e The field error message
 * @returns {Message} The field error to message
 */
export function fieldErrorToMessage(e: FieldError): Message {
  return { message: e.message as string, level: 'error' };
}

/**
 * Phone Options is a list of phones for existing contact
 *
 * @param {IFormData} contact The contact phones to populate number select.
 * @returns {Array} An array of phone nubmers to select.
 */
export const getPhoneOptions = (contact: IFormData) => {
  const items = [];



  if (contact.emailAddresses) {
    items.push({
      key: contact.emailAddresses[0].type,
      value: PhoneValueTypes.MAIL+`: ${contact.emailAddresses[0].address}`,
    });
  }

  if (contact.phoneNumbers) {
    contact.phoneNumbers?.forEach((n: any) => {
      if (n.type === PhoneType.WORK) {
        items.push({
          key: n.type,
          value: PhoneValueTypes.WORK +`: ${n.address}`,
        });
      }
      else if (n.type === PhoneType.MOBILE) {
        items.push({
          key: n.type,
          value: PhoneValueTypes.WORK+`: ${n.address}`,
        });
      }
    });
  }
  return items;
};

/**
 * Handles setting the correct properties on contact when creating new SpeedDial.
 *
 * @param {IFormData} data The speed dial contact
 * @returns {IFormData} Returns contact with correct call types.
 */
export const setPhoneOptions = (data: IFormData) => {
  const contact = { ...data };
  if (!contact.callType) {
    contact.callType = PhoneType.HANDSET;
  }
  
  if (contact.phoneNumbers) {
    contact.phoneNumbers?.forEach((n: any) => {
      if (n.type === PhoneType.WORK) {
        contact.businessPhones = contact.businessPhones || []
        contact.businessPhones.push(n.address);
      }
      else if (n.type === PhoneType.MOBILE) {
        contact.mobilePhone = n.address;
      }
    });
  }
  return contact;
};

/**
 * Get phone types for options select
 *
 * @param {array} types Array of strings
 * @returns {ISelectItems[]} array of select option items.
 */
export const getPhoneTypeOptions = (types: string[]): ISelectItems[] =>
  [...types].map((type: string) => ({
    id: type.toLowerCase(),
    value: type,
  }));

/**
 * Handle abbriviating last name.
 *
 * @param {string} name The display
 * @returns {string} Formatted first and last name.
 */
export function abbrDisplayName(name: string) {
  if (name) {
    const names: string[] = name.split(/\s+/);
    //Added condition to avoid adding undefined if don't have last name
    if (names[1]) {
      names[1] = `${names[1].substr(0, 1)}.`;
    }
    return names.join(' ');
  }

}
