// @ciscospark/widget-roster/src/components/add-participant/index.js
import React from 'react';
import PropTypes from 'prop-types';

import classNames from 'classnames';
import {
  ICON_TYPE_EXIT
} from '@ciscospark/react-component-icon';
import Button from '@ciscospark/react-component-button';

import styles from './styles.css';

export default function AddParticipant(props) {
  function handleChange(e) {
    props.onChange(e.target.value);
  }

  return (
    <div className={classNames(`ciscospark-roster-add-participant`, styles.addPartipicant)}>
      <div className={classNames(styles.searchBar)}>
        <input
          className={classNames(styles.input)}
          onChange={handleChange}
          placeholder={props.placeholder}
          value={props.searchValue}
        />
        <Button
          accessibilityLabel={`Close`}
          buttonClassName={classNames(styles.closeButton)}
          iconType={ICON_TYPE_EXIT}
          onClick={props.onDismiss}
        />
      </div>
    </div>
  );
}

AddParticipant.propTypes = {
  onChange: PropTypes.func.isRequired,
  onDismiss: PropTypes.func.isRequired,
  placeholder: PropTypes.string.isRequired,
  searchValue: PropTypes.string
};
