// @ciscospark/react-component-file-share-display/src/index.js
import React from 'react';
import PropTypes from 'prop-types';

import ShareFile from '@ciscospark/react-component-activity-share-file';
import ShareThumbnail from '@ciscospark/react-component-activity-share-thumbnail';

const defaultProps = {
  isPending: false,
  objectUrl: ''
};

const propTypes = {
  file: PropTypes.object.isRequired,
  isFetching: PropTypes.bool.isRequired,
  isPending: PropTypes.bool,
  objectUrl: PropTypes.string,
  onDownloadClick: PropTypes.func.isRequired
};

function FileShareDisplay({
  file,
  isFetching,
  isPending,
  objectUrl,
  onDownloadClick
}) {
  if (file.image) {
    return (
      <ShareThumbnail
        file={file}
        isFetching={isFetching}
        isPending={isPending}
        objectUrl={objectUrl}
        onDownloadClick={onDownloadClick}
      />
    );
  }
  return (
    <ShareFile
      file={file}
      isPending={isPending}
      onDownloadClick={onDownloadClick}
    />
  );
}

FileShareDisplay.defaultProps = defaultProps;
FileShareDisplay.propTypes = propTypes;

export default FileShareDisplay;
