// @ciscospark/react-component-activity-menu/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import Button from '@ciscospark/react-component-button';
import {ICON_TYPE_EXIT, ICON_TYPE_MESSAGE, ICON_TYPE_VIDEO_OUTLINE} from '@ciscospark/react-component-icon';

import styles from './styles.css';

export default function ActivityMenu(props) {
  const {
    onExit,
    onMeet,
    onMessage
  } = props;

  function handleExit(e) {
    e.preventDefault();
    onExit();
  }

  function handleMeet(e) {
    e.preventDefault();
    onMeet();
  }

  function handleMessage(e) {
    e.preventDefault();
    onMessage();
  }

  return (
    <div className={classNames(`activity-menu`, styles.activityMenu)}>
      <div className={classNames(`exit`, styles.exit)}>
        <div className={classNames(`menu-item`, styles.menuItem)}>
          <div className={classNames(`button-container`, styles.buttonContainer)}>
            <div className={classNames(`button-item`, styles.buttonItem)}>
              <Button iconType={ICON_TYPE_EXIT} onClick={handleExit} title={`Exit`} />
            </div>
          </div>
        </div>
      </div>
      <div className={classNames(`main-menu`, styles.mainMenu)}>
        <div className={classNames(`menu-item`, styles.menuItem)}>
          <div className={classNames(`button-container`, styles.buttonContainer)}>
            <div className={classNames(`button-item`, styles.buttonItem)}>
              <Button iconType={ICON_TYPE_MESSAGE} onClick={handleMessage} title={`Message`} />
            </div>
          </div>
          <div className={classNames(`menu-item-label`, styles.menuItemLabel)}>
            {`Message`}
          </div>
        </div>
        <div className={classNames(`menu-item`, styles.menuItem, `meet`, styles.meet)}>
          <div className={classNames(`button-container`, styles.buttonContainer)}>
            <div className={classNames(`button-item`, styles.buttonItem)}>
              <Button iconType={ICON_TYPE_VIDEO_OUTLINE} onClick={handleMeet} title={`Call`} />
            </div>
          </div>
          <div className={classNames(`menu-item-label`, styles.menuItemLabel)}>
            {`Call`}
          </div>
        </div>
      </div>
    </div>
  );
}

ActivityMenu.propTypes = {
  onExit: PropTypes.func.isRequired,
  onMeet: PropTypes.func.isRequired,
  onMessage: PropTypes.func.isRequired
};
