// @ciscospark/react-component-activity-menu/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import Button from '@ciscospark/react-component-button';
import ButtonControls from '@ciscospark/react-component-button-controls';
import {ICON_TYPE_EXIT, ICON_TYPE_MESSAGE_OUTLINE, ICON_TYPE_VIDEO_OUTLINE} from '@ciscospark/react-component-icon';

import styles from './styles.css';

export default function ActivityMenu(props) {
  const {
    onExit,
    onMeet,
    onMessage
  } = props;

  function handleExit(e) {
    e.preventDefault();
    onExit();
  }

  function handleMeet(e) {
    e.preventDefault();
    onMeet();
  }

  function handleMessage(e) {
    e.preventDefault();
    onMessage();
  }

  const buttons = [
    {
      label: `Message`,
      buttonClassName: styles.messageButton,
      iconType: ICON_TYPE_MESSAGE_OUTLINE,
      onClick: handleMessage
    },
    {
      label: `Call`,
      buttonClassName: styles.meetButton,
      iconType: ICON_TYPE_VIDEO_OUTLINE,
      onClick: handleMeet
    }
  ];

  return (
    <div className={classNames(`activity-menu`, styles.activityMenu)}>
      <div className={classNames(`activity-menu-exit`, styles.menuExit)}>
        <Button
          buttonClassName={styles.exitButton}
          iconType={ICON_TYPE_EXIT}
          onClick={handleExit}
          title={`Exit`}
        />
      </div>
      <div className={classNames(`main-menu`, styles.mainMenu)}>
        <ButtonControls buttons={buttons} showLabels />
      </div>
    </div>
  );
}

ActivityMenu.propTypes = {
  onExit: PropTypes.func.isRequired,
  onMeet: PropTypes.func.isRequired,
  onMessage: PropTypes.func.isRequired
};
