// @webex/private-react-component-token-input/src/index.js
import React, {useState} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Button, Input, Radio, RadioGroup} from '@momentum-ui/react';

import './momentum.scss';
import styles from './styles.css';

const propTypes = {
  onLogin: PropTypes.func.isRequired,
  token: PropTypes.string,
  tokenType: PropTypes.string
};

const defaultProps = {
  token: '',
  tokenType: ''
};

function TokenInput(props) {
  const [tokenSaved, setTokenSaved] = useState(!!props.token);
  const [tokenType, setTokenType] = useState(props.tokenType);
  const [userAccessToken, setUserAccessToken] = useState(props.token);

  function handleClearToken() {
    setUserAccessToken('');
    setTokenSaved(false);
    props.onLogin();
  }

  function handleSaveToken() {
    setTokenSaved(true);
    props.onLogin(userAccessToken, tokenType);
  }

  function handleAccessTokenChange(e) {
    setUserAccessToken(e.target.value);
  }

  function handleTypeChange(value) {
    setTokenType(value);
  }

  return (
    <div className={classNames(styles.section)}>
      <div>
        <h2>Webex Teams Access Token</h2>
        <p>Webex Teams Widgets require an access token to identify the current user.</p>
        {!tokenSaved &&
          <div>
            <div>
              <h3>Access Token Type</h3>
              <RadioGroup
                ariaLabel="Choose Access Token Type"
                name="tokenType"
                onChange={handleTypeChange}
                values={[tokenType]}
              >
                <Radio
                  ariaLabel="Type Access Token"
                  htmlId="tokenTypeAccessToken"
                  label="Access Token"
                  value="token"
                />
                <Radio
                  ariaLabel="Type Guest Token"
                  htmlId="tokenTypeGuestToken"
                  label="Guest Token"
                  value="JWT"
                />
              </RadioGroup>
            </div>
            <div>
              <Input
                aria-label="Access Token"
                htmlId="accessTokenInput"
                inputSize="medium-12"
                label="Token"
                onChange={handleAccessTokenChange}
                placeholder={tokenType === 'JWT' ? 'JWT' : 'Access Token'}
                value={userAccessToken}
              />
              <p>You can get an access token from <a href="http://developer.webex.com">developer.webex.com</a></p>
            </div>
          </div>
        }
        <div>
          {
            !tokenSaved &&
            <Button
              ariaLabel="Save Token"
              color="blue"
              disabled={!userAccessToken}
              id="accessTokenSaveButton"
              onClick={handleSaveToken}
            >
              Save Token
            </Button>
          }
          {
            tokenSaved &&
            <Button
              ariaLabel="Clear Token"
              color="blue"
              onClick={handleClearToken}
            >
              Token Saved, Clear?
            </Button>
          }
        </div>
      </div>
    </div>
  );
}

TokenInput.propTypes = propTypes;
TokenInput.defaultProps = defaultProps;

export default TokenInput;
