// @ciscospark/widget-recents/src/components/space-item/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import styles from './styles.css';

export default function SpaceItem({
  avatarUrl,
  id,
  lastActivityTime,
  lastActivityText,
  name,
  onClick
}) {

  function handleClick() {
    return () => onClick(id);
  }

  return ( // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      className={classNames(`space-item`, styles.item)}
      onClick={handleClick()}
      onKeyDown={handleClick()}
      role="button"
      tabIndex="0"
    >
      <img alt={`${name} Avatar`} src={avatarUrl} />
      <div className={classNames(`space-title`, styles.title)}>{name}</div>
      <div className={classNames(`space-last-activity-time`, styles.timestamp)}>{lastActivityTime}</div>
      <div className={classNames(`space-last-activity`, styles.lastActivity)}>{lastActivityText}</div>
    </div>
  );
}

SpaceItem.propTypes = {
  avatarUrl: PropTypes.string,
  id: PropTypes.string,
  lastActivityText: PropTypes.string,
  lastActivityTime: PropTypes.string,
  name: PropTypes.string,
  onClick: PropTypes.func
};
