// @ciscospark/widget-recents/src/components/space-item/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import Avatar from '@ciscospark/react-component-avatar';

import styles from './styles.css';

export default function SpaceItem({
  avatarUrl,
  id,
  isUnread,
  lastActivityTime,
  lastActivityText,
  name,
  onClick
}) {

  function handleClick() {
    return () => onClick(id);
  }


  return ( // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      className={classNames(`space-item`, styles.item)}
      onClick={handleClick()}
      onKeyDown={handleClick()}
      role="button"
      tabIndex="0"
    >
      {
        isUnread &&
        <div className={classNames(`space-unread-indicator`, styles.unreadIndicator)} />
      }
      <div className={classNames(`space-avatar-wrapper`, styles.avatarWrapper)}>
        <Avatar name={name} />
      </div>
      <div className={classNames(`space-item-meta`, styles.meta)}>
        <div className={classNames(`space-title`, styles.title)}>{name}</div>
        <div className={classNames(`space-last-activity`, styles.lastActivity)}>{lastActivityText}</div>
      </div>
      <div className={classNames(`space-last-activity-time`, styles.timestamp)}>{lastActivityTime}</div>
    </div>
  );
}

SpaceItem.propTypes = {
  avatarUrl: PropTypes.string,
  id: PropTypes.string,
  isUnread: PropTypes.bool,
  lastActivityText: PropTypes.string,
  lastActivityTime: PropTypes.string,
  name: PropTypes.string,
  onClick: PropTypes.func
};
