// @ciscospark/react-container-call-incoming/src/index.js
import React, {Component, PropTypes} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';

import Ringtone, {RINGTONE_TYPE_INCOMING} from '@ciscospark/react-component-ringtone';

import {
  callIncoming,
  listenForIncomingCalls,
  removeCallIncoming
} from '@ciscospark/redux-module-media';

import {
  NOTIFICATION_TYPE_CALL,
  createNotification
} from '@ciscospark/react-container-notifications';


export class IncomingCall extends Component {

  componentDidMount() {
    const {media, spark, onEvent} = this.props;
    if (media.getIn([`status`, `isListening`])) {
      return;
    }
    spark.phone.on(`call:incoming`, (incomingCall) => {
      if (!this.shouldAcknowledgeCall(incomingCall)) {
        return;
      }
      incomingCall.acknowledge()
        .then(() => {
          this.props.callIncoming(incomingCall);
          const removeIncoming = () => this.props.removeCallIncoming(incomingCall);

          // Disconnects will happen if the remote party hangs up before we accept/decline
          incomingCall.once(`disconnected`, removeIncoming);
          // Once we are connected, our connected call state will handle disconnects
          incomingCall.once(`connected`, () => incomingCall.off(`disconnected`, removeIncoming));

          // Emit the event
          onEvent(`calls:ringing`, {
            actor: incomingCall.from.person,
            call: incomingCall
          });

          // Create notification of call
          this.props.createNotification(incomingCall.locusUrl, NOTIFICATION_TYPE_CALL);
        });
    });
    this.props.listenForIncomingCalls();
  }

  shouldAcknowledgeCall(incomingCall) {
    const {media, conversation} = this.props;

    // We only acknowledge calls directly related to this widget
    if (incomingCall.locus.url !== conversation.locusUrl) {
      return false;
    }
    // If we make an outbound call via the widget, when the remote party answers,
    // we get a call:incoming notification when they connect
    if (media.getIn([`status`, `isDialing`])) {
      return false;
    }
    return true;
  }

  render() {
    const isRinging = this.props.media.getIn([`status`, `isRinging`]);
    return (
      <Ringtone
        play={isRinging}
        type={RINGTONE_TYPE_INCOMING}
      />
    );
  }
}

function mapStateToProps(state) {
  return {
    conversation: state.conversation,
    media: state.media,
    spark: state.spark.get(`spark`)
  };
}

const injectedPropTypes = {
  conversation: PropTypes.object,
  media: PropTypes.object,
  spark: PropTypes.object,
  callIncoming: PropTypes.func,
  createNotification: PropTypes.func,
  listenForIncomingCalls: PropTypes.func,
  removeCallIncoming: PropTypes.func
};

IncomingCall.propTypes = {
  onEvent: PropTypes.func,
  ...injectedPropTypes
};

export default connect(
  mapStateToProps,
  (dispatch) => bindActionCreators({
    callIncoming,
    createNotification,
    removeCallIncoming,
    listenForIncomingCalls
  }, dispatch)
)(IncomingCall);
