// @ciscospark/react-container-call-incoming/src/index.js
import {Component, PropTypes} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {Howl} from 'howler';

import {
  callIncoming,
  listeningForIncomingCalls,
  removeCallIncoming
} from '@ciscospark/redux-module-media';

import {
  NOTIFICATION_TYPE_CALL,
  createNotification
} from '@ciscospark/react-container-notifications';

import Ringtone from './media/Ringtone.mp3';


export class IncomingCall extends Component {
  constructor(props) {
    super(props);
    this.ringtone = new Howl({
      src: [Ringtone],
      loop: true
    });
  }

  componentDidMount() {
    const {media, spark, onEvent} = this.props;
    if (media.getIn([`status`, `isListening`])) {
      return;
    }
    spark.phone.on(`call:incoming`, (incomingCall) => {
      if (!this.shouldAcknowledgeCall(incomingCall)) {
        return;
      }
      incomingCall.acknowledge()
        .then(() => {
          this.props.callIncoming(incomingCall);
          const removeIncoming = () => this.props.removeCallIncoming(incomingCall);

          // Disconnects will happen if the remote party hangs up before we accept/decline
          incomingCall.once(`disconnected`, removeIncoming);
          // Once we are connected, our connected call state will handle disconnects
          incomingCall.once(`connected`, () => incomingCall.off(`disconnected`, removeIncoming));

          // Emit the event
          onEvent(`calls:ringing`, {
            actor: incomingCall.from.person,
            call: incomingCall
          });

          // Create notification of call
          this.props.createNotification(incomingCall.locusUrl, NOTIFICATION_TYPE_CALL);
        });
    });
    this.props.listeningForIncomingCalls();
  }

  componentWillReceiveProps(nextProps) {
    const isRinging = nextProps.media.getIn([`status`, `isRinging`]);
    const wasRinging = this.props.media.getIn([`status`, `isRinging`]);
    // Change in ringing state
    if (isRinging !== wasRinging) {
      const isPlaying = this.ringtone.playing();
      if (isRinging && !isPlaying) {
        this.ringtone.play();
      }
      if (!isRinging && isPlaying) {
        this.ringtone.stop();
      }
    }
  }


  shouldAcknowledgeCall(incomingCall) {
    const {media, conversation} = this.props;

    // We only acknowledge calls directly related to this widget
    if (incomingCall.locus.url !== conversation.locusUrl) {
      return false;
    }
    // If we make an outbound call via the widget, when the remote party answers,
    // we get a call:incoming notification when they connect
    if (media.getIn([`status`, `isDialing`])) {
      return false;
    }
    return true;
  }

  render() {
    return null;
  }
}

function mapStateToProps(state) {
  return {
    conversation: state.conversation,
    media: state.media,
    spark: state.spark.get(`spark`)
  };
}

const injectedPropTypes = {
  conversation: PropTypes.object,
  media: PropTypes.object,
  spark: PropTypes.object,
  callIncoming: PropTypes.func,
  createNotification: PropTypes.func,
  listeningForIncomingCalls: PropTypes.func,
  removeCallIncoming: PropTypes.func
};

IncomingCall.propTypes = {
  onEvent: PropTypes.func,
  ...injectedPropTypes
};

export default connect(
  mapStateToProps,
  (dispatch) => bindActionCreators({
    callIncoming,
    createNotification,
    removeCallIncoming,
    listeningForIncomingCalls
  }, dispatch)
)(IncomingCall);
