// @ciscospark/widget-message/src/container.js
import React, {Component, PropTypes} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {intlShape, injectIntl} from 'react-intl';
import Dropzone from 'react-dropzone';
import classNames from 'classnames';
import _ from 'lodash';

import {constructFiles} from '@ciscospark/react-component-utils';
import {fetchAvatarForUserId} from '@ciscospark/redux-module-avatar';
import {
  acknowledgeActivityOnServer,
  createConversationWithUser,
  loadPreviousMessages,
  receiveMercuryActivity,
  receiveMercuryComment,
  updateMercuryState,
  VISIBLE_ACTIVITY_VERBS,
  VISIBLE_ACTIVITY_TYPES
} from '@ciscospark/redux-module-conversation';
import {
  fetchFlags,
  flagActivity,
  removeFlagFromServer
} from '@ciscospark/redux-module-flags';
import {
  setTyping
} from '@ciscospark/redux-module-indicators';
import {addFiles, removeInflightActivity} from '@ciscospark/redux-module-activity';
import ConfirmationModal from '@ciscospark/react-component-confirmation-modal';
import Cover from '@ciscospark/react-component-cover';
import LoadingScreen from '@ciscospark/react-component-loading-screen';
import ScrollToBottomButton from '@ciscospark/react-component-scroll-to-bottom-button';
import ActivityList from '@ciscospark/react-container-activity-list';
import MessageComposer from '@ciscospark/react-container-message-composer';
import ReadReceipts from '@ciscospark/react-container-read-receipts';
import ScrollingActivity from '@ciscospark/react-container-scrolling-activity';

import {
  NOTIFICATION_TYPE_POST,
  createNotification
} from '@ciscospark/react-container-notifications';

import {
  confirmDeleteActivity,
  deleteActivityAndDismiss,
  hideDeleteModal,
  setScrollPosition,
  showScrollToBottomButton,
  updateHasNewMessage
} from './actions';

import styles from './styles.css';
import messages from './messages';


/**
 * MessageWidget Container
 * @extends Component
 */
export class MessageWidget extends Component {

  constructor(props) {
    super(props);
    this.getActivityList = this.getActivityList.bind(this);
    this.handleActivityDelete = this.handleActivityDelete.bind(this);
    this.handleActivityFlag = this.handleActivityFlag.bind(this);
    this.handleConfirmActivityDelete = this.handleConfirmActivityDelete.bind(this);
    this.handleCancelActivityDelete = this.handleCancelActivityDelete.bind(this);
    this.handleFileDrop = this.handleFileDrop.bind(this);
    this.handleMessageSubmit = this.handleMessageSubmit.bind(this);
    this.handleScroll = _.debounce(this.handleScroll.bind(this), 150);
  }

  componentDidMount() {
    // If our widget is instantiated with all the props loaded,
    // we can setup the conversation
    this.establishConversation(this.props);
  }

  componentWillReceiveProps(nextProps) {
    // This needs to fire here because props are loaded async via redux
    this.establishConversation(nextProps);
  }

  shouldComponentUpdate(nextProps) {
    const props = this.props;

    return nextProps.conversation !== props.conversation
      || nextProps.flags !== props.flags
      || nextProps.indicators !== props.indicators
      || nextProps.share !== props.share
      || nextProps.user !== props.user
      || nextProps.widget !== props.widget
      || nextProps.activity !== props.activity;
  }

  componentWillUpdate(nextProps) {
    const props = this.props;
    if (this.activityList && nextProps.conversation.activities.size !== props.conversation.activities.size) {
      this.scrollHeight = this.activityList.getScrollHeight();
    }
  }

  componentDidUpdate(prevProps) {
    const props = this.props;
    const activityList = this.activityList;
    if (activityList) {
      const activities = props.conversation.activities;
      const prevActivities = prevProps.conversation.activities;
      const lastActivityFromPrev = prevActivities.last();
      const lastActivityFromThis = activities.last();

      // If new activity comes in
      if (lastActivityFromPrev
        && lastActivityFromThis
        && activities.size !== prevActivities.size
        && lastActivityFromPrev.id !== lastActivityFromThis.id) {
        if (props.user.get(`currentUser`).id !== lastActivityFromThis.actor.id) {
          // Send notification of new message
          props.createNotification(lastActivityFromThis.url, NOTIFICATION_TYPE_POST);
        }
      }
      this.updateScroll(activities, prevActivities, prevProps.widget.scrollTop);
      if (this.shouldScrollToBottom(props, prevProps)) {
        activityList.scrollToBottom();
      }
    }
  }

  /**
   * Check if activity list should scroll to bottom
   *
   * @param {object} props
   * @param {object} prevProps
   * @returns {boolean}
   */
  shouldScrollToBottom(props, prevProps) {
    let shouldScrollToBottom = false;
    const activityList = this.activityList;
    const lastActivityFromPrev = prevProps.conversation.activities.last();
    const lastActivityFromThis = props.conversation.activities.last();
    // If new activity comes in
    if (lastActivityFromPrev && lastActivityFromThis && props.conversation.activities.size !== prevProps.conversation.activities.size && lastActivityFromPrev.id !== lastActivityFromThis.id) {
      // Scroll if from ourselves
      if (props.user.get(`currentUser`).id === lastActivityFromThis.actor.id) {
        shouldScrollToBottom = true;
      }
      else if (activityList.isScrolledToBottom()) {
        shouldScrollToBottom = true;
      }
    }
    else if (prevProps.conversation.activities.size === 0) {
      shouldScrollToBottom = true;
    }
    // Scroll to show in flight activities
    if (props.activity.get(`inFlightActivities`).size && props.activity.get(`inFlightActivities`).size !== prevProps.activity.get(`inFlightActivities`).size) {
      shouldScrollToBottom = true;
    }

    return shouldScrollToBottom;
  }

  /**
   * Update scroll position when activity list changes
   * @param {object} activities
   * @param {object} prevActivities
   * @param {number} prevScrollTop
   * @returns {null}
   */
  updateScroll(activities, prevActivities, prevScrollTop) {
    const firstActivityFromPrev = prevActivities.first();
    const firstActivityFromThis = activities.first();
    const activityList = this.activityList;

    if (firstActivityFromThis && firstActivityFromPrev && firstActivityFromThis.id !== firstActivityFromPrev.id) {
      activityList.setScrollTop(activityList.getScrollHeight() - this.scrollHeight + prevScrollTop);
    }
  }


  /**
   * Attempts to fetch and establish a conversation
   *
   * @param {object} props
   *
   * @memberOf MessageWidget
   *
   * @returns {null}
   */
  establishConversation(props) {
    const {
      toPersonEmail,
      toPersonId,
      spark,
      sparkState,
      conversation
    } = props;

    const {
      authenticated,
      connected,
      registered
    } = sparkState;

    // Check if Spark instance and connections have been established
    if (spark && connected && authenticated && registered) {
      // Check if conversation has been retrieved
      if (!conversation.id && !conversation.isFetching) {
        props.createConversationWithUser(toPersonId || toPersonEmail, spark);
      }
    }
    // Setup once we have a conversation
    if (conversation.id) {
      this.setupConversationActions(conversation, props);
    }
  }

  /**
   * Once a conversation has been established, setup
   * actions need to happen for app state
   *
   * @param {any} conversation
   * @param {any} nextProps
   *
   * @returns {null}
   */
  setupConversationActions(conversation, nextProps) {
    const {flags, spark} = nextProps;
    const {props} = this;
    const prevConversation = props.conversation;
    if (!conversation.mercuryState.isListening) {
      this.listenToTypingEvents(conversation.id, spark);
      this.listenToNewActivity(conversation.id, spark);
    }
    if (!flags.hasFetched && !flags.isFetching) {
      nextProps.fetchFlags(spark);
    }
    // We only want to fetch avatars when a new activity is seen
    if (conversation.participants.length && conversation.participants.length !== prevConversation.participants.length) {
      this.getAvatarsFromConversation(conversation);
    }
  }

  /**
   * Setup listeners for typing events
   *
   * @param {String} conversationId
   * @param {Object} spark
   * @returns {undefined}
   */
  listenToTypingEvents(conversationId, spark) {
    const {props} = this;
    props.updateMercuryState({isListening: true});
    spark.mercury.on(`event:status.start_typing`, (event) => {
      if (event.data.conversationId === conversationId) {
        props.setTyping(event.data.actor.id, true);
      }
    });

    spark.mercury.on(`event:status.stop_typing`, (event) => {
      if (event.data.conversationId === conversationId) {
        props.setTyping(event.data.actor.id, false);
      }
    });
  }

  /**
   * Setup listeners for new activities
   *
   * @param {String} conversationId
   * @param {Object} spark
   * @returns {undefined}
   */
  listenToNewActivity(conversationId, spark) {
    const {props} = this;
    spark.mercury.on(`event:conversation.activity`, (event) => {
      const activity = event.data.activity;
      const isVisibleContent = VISIBLE_ACTIVITY_VERBS.indexOf(activity.verb) !== -1
        && VISIBLE_ACTIVITY_TYPES.indexOf(activity.object.objectType) !== -1;
      // Ignore activity from other conversations
      if (activity.target.id === conversationId) {
        if (isVisibleContent) {
          props.updateHasNewMessage(true);
          props.receiveMercuryComment(activity);
          props.removeInflightActivity(activity.clientTempId);
        }
        else if (activity.object.objectType === `activity`) {
          props.receiveMercuryActivity(activity);
        }
      }
    });
  }

  /**
   * Scroll activityList to bottom
   *
   * @returns {undefined}
   */
  handleScrollToBottom() {
    this.activityList.scrollToBottom();
  }

  /**
   * Additional actions after submitting a message
   *
   * @returns {undefined}
   */
  handleMessageSubmit() {
    this.activityList.scrollToBottom();
  }

  /**
   * Processes the activities and fetches avatars for users
   * that have not been fetched yet
   *
   * @param {Object} conversation
   *
   * @returns {null}
   */
  getAvatarsFromConversation(conversation) {
    const {props} = this;
    const {
      avatar,
      spark
    } = props;
    const {participants} = conversation;
    const userIds = _.uniq(participants.map((participant) => participant.id))
      .filter((userId) => {
        // Only return users that haven't been fetched or is fetching
        const fetched = avatar.hasIn([`items`, userId]);
        const fetching = avatar.hasIn([`avatarsInFlight`, userId]);
        return !fetched || !fetching;
      });
    userIds.forEach((userId) => props.fetchAvatarForUserId(userId, spark));
  }

  /**
   * Store activity list from child component
   *
   * @param {Object} ref
   * @returns {undefined}
   */
  getActivityList(ref) {
    this.activityList = ref;
  }

  /**
   * Flag or unflag activity by Id
   *
   * @param {String} activityId
   * @returns {undefined}
   */
  handleActivityFlag(activityId) {
    const props = this.props;
    const {
      conversation,
      flags,
      spark
    } = props;
    const activity = conversation.activities.find((act) => act.id === activityId);
    if (activity) {
      const foundFlag = flags.flags.find((flag) => flag.activityUrl === activity.url);
      if (foundFlag) {
        props.removeFlagFromServer(foundFlag, spark);
      }
      else {
        props.flagActivity(activity, spark);
      }

    }
  }

  /**
   * Displays modal confirming activity delete
   *
   * @param {String} activityId
   * @returns {undefined}
   */
  handleActivityDelete(activityId) {
    const props = this.props;
    props.confirmDeleteActivity(activityId);
  }

  /**
   * Does the actual deletion of the activity after confirmation modal
   *
   * @returns {undefined}
   */
  handleConfirmActivityDelete() {
    const {props} = this;
    const {
      conversation,
      spark,
      widget
    } = props;
    const activityId = widget.deletingActivityId;
    const activity = conversation.activities.find((act) => act.id === activityId);
    if (activity) {
      props.deleteActivityAndDismiss(conversation, activity, spark);
    }
    else {
      props.hideDeleteModal();
    }
  }


  /**
   * Dismisses the confirmation modal
   *
   * @returns {undefined}
   */
  handleCancelActivityDelete() {
    const props = this.props;
    props.hideDeleteModal();
  }

  /**
   * Adds dropped file into activity store
   *
   * @param {array} acceptedFiles
   * @returns {undefined}
   */
  handleFileDrop(acceptedFiles) {
    const props = this.props;
    const {
      activity,
      conversation,
      spark
    } = props;
    const files = constructFiles(acceptedFiles);
    props.addFiles(conversation, activity, files, spark);
  }

  /**
   * Perform actions when activity list scrolls
   *
   * @returns {undefined}
   */
  handleScroll() {
    const props = this.props;
    const activityList = this.activityList;
    const {
      conversation,
      spark,
      widget
    } = props;

    props.setScrollPosition({scrollTop: activityList.getScrollTop()});

    const lastActivity = conversation.activities.last();
    if (activityList.isScrolledToBottom()) {
      props.showScrollToBottomButton(false);
      props.updateHasNewMessage(false);
      if (conversation.lastAcknowledgedActivityId !== lastActivity.id) {
        props.acknowledgeActivityOnServer(conversation, lastActivity, spark);
      }
    }
    else if (!widget.showScrollToBottomButton) {
      props.showScrollToBottomButton(true);
    }

    if (activityList.isScrolledToTop() && conversation.activities.first().verb !== `create`) {
      props.loadPreviousMessages(conversation.id, conversation.activities.first().published, spark);
    }
  }

  /**
   * Gets the non-current user of a conversation
   *
   * @param {object} conversation
   * @returns {object}
   */
  getToUserFromConversation(conversation) {
    if (!conversation.participants) {
      return null;
    }
    const props = this.props;
    return conversation.participants.find((user) =>
      user.id !== props.user.get(`currentUser`).id
    );
  }

  /**
   * Renders the conversation area of the widget
   *
   * @returns {object}
   */
  render() {
    const props = this.props;
    const {
      conversation,
      spark,
      widget
    } = props;
    const {formatMessage} = this.props.intl;

    const {
      isLoadingHistoryUp
    } = conversation;

    const toUser = this.getToUserFromConversation(conversation);

    if (conversation && conversation.id) {
      return (
        <Dropzone
          activeClassName={styles.activeDropzone}
          className={styles.dropzone} // eslint-disable-line react/forbid-component-props
          disableClick
          disablePreview
          onDrop={this.handleFileDrop}
        >
          <div className={classNames(`activity-list-wrapper`, styles.activityListWrapper)}>
            <ScrollingActivity
              isLoadingHistoryUp={isLoadingHistoryUp}
              onScroll={this.handleScroll}
              ref={this.getActivityList}
            >
              <ActivityList
                onActivityDelete={this.handleActivityDelete}
                onActivityFlag={this.handleActivityFlag}
              />
              <div className={classNames(`indicators`, styles.indicators)}><ReadReceipts /></div>
            </ScrollingActivity>
            {
              widget.showScrollToBottomButton &&
              <ScrollToBottomButton
                label={formatMessage(messages.newMessagesMessage)}
                onClick={this.handleScrollToBottom}
              />
            }
          </div>
          <div className={classNames(`message-composer-wrapper`, styles.messageComposerWrapper)}>
            <MessageComposer
              conversation={conversation}
              onSubmit={this.handleMessageSubmit}
              placeholder={formatMessage(messages.messageComposerPlaceholder, {displayName: toUser.displayName})}
              spark={spark}
            />
          </div>
          {
            props.widget.showAlertModal &&
            <ConfirmationModal
              actionButtonLabel={formatMessage(messages.deleteButtonLabel)}
              body={formatMessage(messages.deleteAlertBody)}
              cancelButtonLabel={formatMessage(messages.cancelButtonLabel)}
              onClickActionButton={this.handleConfirmActivityDelete}
              onClickCancelButton={this.handleCancelActivityDelete}
              title={formatMessage(messages.deleteAlertTitle)}
            />
          }
          <Cover message={formatMessage(messages.dropzoneCoverMessage)} />
        </Dropzone>
      );
    }
    return <LoadingScreen />;
  }
}

MessageWidget.propTypes = {
  conversation: PropTypes.object,
  intl: intlShape.isRequired,
  spark: PropTypes.object.isRequired,
  toPersonEmail: PropTypes.string,
  toPersonId: PropTypes.string
};

function mapStateToProps(state, ownProps) {
  const {
    avatar,
    activity,
    conversation,
    flags,
    spark,
    user,
    widget
  } = state;
  return {
    avatar,
    activity: ownProps.activity || activity,
    conversation: ownProps.conversation || conversation,
    flags,
    spark: ownProps.spark || spark,
    sparkState: state.spark.get(`status`).toJS(),
    user,
    widget
  };
}

export default connect(
  mapStateToProps,
  (dispatch) => bindActionCreators({
    acknowledgeActivityOnServer,
    addFiles,
    confirmDeleteActivity,
    createConversationWithUser,
    createNotification,
    deleteActivityAndDismiss,
    fetchAvatarForUserId,
    fetchFlags,
    flagActivity,
    hideDeleteModal,
    loadPreviousMessages,
    removeFlagFromServer,
    receiveMercuryActivity,
    receiveMercuryComment,
    removeInflightActivity,
    setScrollPosition,
    setTyping,
    showScrollToBottomButton,
    updateHasNewMessage,
    updateMercuryState
  }, dispatch)
)(injectIntl(MessageWidget));
