// @ciscospark/widget-message/src/container.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import Dropzone from 'react-dropzone';
import classNames from 'classnames';
import {debounce, has, throttle} from 'lodash';
import {autobind} from 'core-decorators';

import {constructFiles} from '@ciscospark/react-component-utils';

import {fetchAvatarsForUsers} from '@ciscospark/redux-module-avatar';
import {
  getFeature,
  FEATURE_MENTIONS
} from '@ciscospark/redux-module-features';

import {
  acknowledgeActivityOnServer,
  createConversation,
  getConversation,
  loadMissingActivities,
  loadPreviousMessages
} from '@ciscospark/redux-module-conversation';

import {
  fetchFlags,
  flagActivity,
  removeFlagFromServer
} from '@ciscospark/redux-module-flags';

import {setTyping} from '@ciscospark/redux-module-indicators';

import {
  addFiles,
  removeInflightActivity,
  retryFailedActivity
} from '@ciscospark/redux-module-activity';

import {
  subscribeToPresenceUpdates,
  unsubscribeFromPresenceUpdates
} from '@ciscospark/redux-module-presence';


import ConfirmationModal from '@ciscospark/react-component-confirmation-modal';
import Cover from '@ciscospark/react-component-cover';
import LoadingScreen from '@ciscospark/react-component-loading-screen';
import ScrollToBottomButton from '@ciscospark/react-component-scroll-to-bottom-button';
import ActivityList from '@ciscospark/react-container-activity-list';
import MessageComposer from '@ciscospark/react-container-message-composer';
import ReadReceipts from '@ciscospark/react-container-read-receipts';
import ScrollingActivity from '@ciscospark/react-container-scrolling-activity';
import ErrorDisplay from '@ciscospark/react-component-error-display';

import wrapConversationMercury from '@ciscospark/react-hoc-conversation-mercury';

import Notifications, {
  NOTIFICATION_TYPE_POST,
  createNotification
} from '@ciscospark/react-container-notifications';

import {
  confirmDeleteActivity,
  deleteActivityAndDismiss,
  hideDeleteModal,
  setScrolledUp,
  setScrollPosition,
  showScrollToBottomButton,
  updateHasNewMessage,
  updateWidgetState
} from './actions';

import styles from './styles.css';
import messages from './messages';

import getMessageWidgetProps from './selector';
import {eventNames as defaultEventNames} from './events';

import {handleConversationActivityEvent} from './helpers';

/**
 * MessageWidget Container
 * @extends Component
 */
export class MessageWidget extends Component {
  /**
   * Processes the activities and fetches avatars for users
   * that have not been fetched yet
   *
   * @param {Object} props
   *
   */
  static getAvatarsFromParticipants(props) {
    const {
      participants,
      sparkInstance
    } = props;
    if (participants.length === 0) {
      return;
    }
    const userIds = participants.map((p) => p.id);
    props.fetchAvatarsForUsers(userIds, sparkInstance);
  }

  /**
   * Check if activity list should scroll to bottom
   *
   * @param {object} props
   * @param {object} prevProps
   * @returns {boolean}
   */
  static shouldScrollToBottom(props, prevProps) {
    const {
      activityCount
    } = props;
    const prevActivitiesCount = prevProps.activityCount;
    if (activityCount && prevActivitiesCount === 0) {
      // Always scroll to the bottom when activities first load
      return true;
    }

    // Otherwise, don't scroll if we have scrolled up
    let shouldScrollToBottom = true;
    if (props.widgetMessage.get('hasScrolledUp')) {
      shouldScrollToBottom = false;
    }
    return shouldScrollToBottom;
  }

  constructor(props) {
    super(props);
    this.handleScroll = throttle(this.handleScroll, 250);
    this.handleMouseMove = debounce(this.handleMouseMove, 500, {
      leading: true,
      trailing: false
    });
  }

  componentDidMount() {
    this.eventNames = Object.assign({}, defaultEventNames, this.props.eventNames);
    // If our widget is instantiated with all the props loaded,
    // we can setup the conversation
    this.establishConversation(this.props);
  }

  componentWillReceiveProps(nextProps) {
    // This needs to fire here because props are loaded async via redux
    this.establishConversation(nextProps);
  }

  shouldComponentUpdate(nextProps) {
    const {props} = this;

    return nextProps.conversation !== props.conversation
      || nextProps.flags !== props.flags
      || nextProps.indicators !== props.indicators
      || nextProps.share !== props.share
      || nextProps.user !== props.user
      || nextProps.widgetMessage !== props.widgetMessage
      || nextProps.activity !== props.activity;
  }

  componentWillUpdate(nextProps) {
    const {props} = this;

    if (this.activityList
      && nextProps.activityCount !== props.activityCount) {
      this.scrollHeight = this.activityList.getScrollHeight();
    }

    if (!nextProps.features.hasIn(['items', FEATURE_MENTIONS])) {
      props.getFeature('developer', FEATURE_MENTIONS, nextProps.sparkInstance);
    }
  }

  componentDidUpdate(prevProps) {
    const {
      props,
      activityList
    } = this;

    if (activityList) {
      const {
        activityCount,
        firstActivity
      } = props;

      const prevActivitiesCount = prevProps.activityCount;
      const previousFirstActivity = prevProps.firstActivity;
      const lastActivityFromPrev = prevProps.lastActivity;
      const lastActivityFromThis = props.lastActivity;
      const objectType = lastActivityFromThis.object ? lastActivityFromThis.object.objectType : undefined;

      // If new activity comes in
      if (lastActivityFromPrev
        && lastActivityFromThis
        && activityCount !== prevActivitiesCount
        && lastActivityFromPrev.id !== lastActivityFromThis.id
        && props.currentUser.id !== lastActivityFromThis.actor.id
        && objectType !== 'locusSessionSummary'
      ) {
        // Send notification of new message
        props.createNotification(lastActivityFromThis.url, NOTIFICATION_TYPE_POST);
      }
      this.updateScroll(firstActivity, previousFirstActivity, prevProps.widgetMessage.get('scrollPosition').scrollTop);

      if (MessageWidget.shouldScrollToBottom(props, prevProps)) {
        activityList.scrollToBottom();
      }
    }
  }

  componentWillUnmount() {
    this.props.unsubscribeFromPresenceUpdates(this.props.participants.map((p) => p.id), this.props.sparkInstance);
  }


  /**
   * Gets the non-current user of a conversation
   *
   * @param {object} conversation
   * @returns {object}
   */
  @autobind
  getToUserFromConversation(conversation) {
    const {props} = this;
    if (!conversation.get('participants') || !props.currentUser) {
      return null;
    }
    return conversation.get('participants').find((user) =>
      user.get('id') !== props.currentUser.id);
  }


  /**
   * Store activity list from child component
   *
   * @param {Object} ref
   * @returns {undefined}
   */
  @autobind
  getActivityList(ref) {
    this.activityList = ref;
  }

  /**
   * Once a conversation has been established, setup
   * actions need to happen for app state
   *
   * @param {any} conversation
   * @param {any} nextProps
   *
   */
  setupConversationActions(conversation, nextProps) {
    const {
      flags, sparkInstance, space, widgetMessage, conversationId
    } = nextProps;
    const {props} = this;
    if (!widgetMessage.get('isListeningToActivity')) {
      this.listenToNewActivity(space, sparkInstance);
    }
    if (!widgetMessage.get('isListeningToTyping')) {
      this.listenToTypingEvents(conversationId, sparkInstance);
    }
    if (!flags.getIn(['status', 'hasFetched']) && !flags.getIn(['status', 'isFetching'])) {
      nextProps.fetchFlags(sparkInstance);
    }
    // We only want to fetch avatars when a new activity is seen
    const {activityCount} = nextProps;
    const prevActivitiesCount = props.activityCount;
    if (activityCount && activityCount !== prevActivitiesCount) {
      MessageWidget.getAvatarsFromParticipants(nextProps);
    }
    if (nextProps.participants) {
      const userIds = nextProps.participants.map((p) => p.id);
      this.props.subscribeToPresenceUpdates(userIds, sparkInstance);
    }
    if (!widgetMessage.get('isListeningToBufferState')) {
      this.listenToBufferState(conversationId, sparkInstance);
    }
  }


  /**
   * Update scroll position when activity list changes
   * @param {object} firstActivity
   * @param {object} previousFirstActivity
   * @param {number} prevScrollTop
   */
  updateScroll(firstActivity, previousFirstActivity, prevScrollTop) {
    const {activityList} = this;

    if (firstActivity && previousFirstActivity && firstActivity.id !== previousFirstActivity.id) {
      activityList.setScrollTop(activityList.getScrollHeight() - this.scrollHeight + prevScrollTop);
    }
  }

  /**
   * Attempts to fetch and establish a conversation
   *
   * @param {object} props
   *
   * @memberOf MessageWidget
   *
   */
  establishConversation(props) {
    const {
      toPersonEmail,
      toPersonId,
      spaceId,
      sparkInstance,
      sparkState,
      conversation,
      conversationId
    } = props;

    const {
      authenticated,
      registered
    } = sparkState;

    // Check if Spark instance and connections have been established
    if (sparkInstance && authenticated) {
      // Check if conversation has been retrieved
      if (registered && !conversationId && !conversation.getIn(['status', 'isFetching'])) {
        if (toPersonId || toPersonEmail) {
          props.createConversation([toPersonId || toPersonEmail], sparkInstance);
        }
        else if (spaceId) {
          props.getConversation(spaceId, sparkInstance);
        }
      }
    }
    // Setup once we have a conversation
    if (conversationId) {
      this.setupConversationActions(conversation, props);
    }
  }

  /**
   * Setup listeners for typing events
   *
   * @param {String} conversationId
   * @param {Object} sparkInstance
   * @returns {undefined}
   */
  listenToTypingEvents(conversationId, sparkInstance) {
    const {props} = this;
    props.updateWidgetState({isListeningToTyping: true});
    sparkInstance.internal.mercury.on('event:status.start_typing', (event) => {
      if (event.data.conversationId === conversationId) {
        props.setTyping(event.data.actor.id, true);
      }
    });

    sparkInstance.internal.mercury.on('event:status.stop_typing', (event) => {
      if (event.data.conversationId === conversationId) {
        props.setTyping(event.data.actor.id, false);
      }
    });
  }

  /**
   * Setup listeners for new activities
   *
   * @param {String} space
   * @param {Object} sparkInstance
   * @returns {undefined}
   */
  listenToNewActivity(space, sparkInstance) {
    const {props} = this;
    const {handleEvent} = this;
    const {currentUser} = props;
    if (currentUser) {
      props.updateWidgetState({isListeningToActivity: true});
      sparkInstance.internal.mercury.on('event:conversation.activity', (event) => {
        handleConversationActivityEvent(event, this.eventNames, currentUser.id, space, {
          handleEvent,
          removeInflightActivity: props.removeInflightActivity,
          updateHasNewMessage: props.updateHasNewMessage
        });
      });
    }
  }

  /**
   * Listens for mercury buffer state events and
   * fires off actions based on them
   * @param {string} conversationId
   * @param {object} sparkInstance
   * @returns {undefined}
   */
  @autobind
  listenToBufferState(conversationId, sparkInstance) {
    const {props} = this;
    props.updateWidgetState({isListeningToBufferState: true});

    sparkInstance.internal.mercury.on('event:mercury.buffer_state', (event) => {
      const {
        lastActivity
      } = props;
      if (has(event, 'data.bufferState.conversation') && event.data.bufferState.conversation === 'UNKNOWN') {
        // Mercury does not contain enough information about the conversation, refetch
        const sinceDate = lastActivity.published || null;
        props.loadMissingActivities(conversationId, sinceDate, sparkInstance);
      }
    });
  }

  @autobind
  acknowledgeLastActivity() {
    const {
      props
    } = this;
    const {
      conversation,
      hasAcknowledgementsDisabled,
      lastActivity,
      sparkInstance
    } = props;
    if (conversation.get('lastAcknowledgedActivityId') !== lastActivity.id) {
      if (!hasAcknowledgementsDisabled) {
        props.acknowledgeActivityOnServer(conversation, lastActivity, sparkInstance);
      }
    }
  }

  /**
   * Scroll activityList to bottom
   *
   * @returns {undefined}
   */
  @autobind
  handleScrollToBottom() {
    this.activityList.scrollToBottom();
  }

  /**
   * Additional actions after submitting a message
   *
   * @returns {undefined}
   */
  @autobind
  handleMessageSubmit() {
    this.activityList.scrollToBottom();
  }

  /**
   * Event handler in case one isn't provided
   * @param {string} name
   * @param {object} data
   */
  @autobind
  handleEvent(name, data) {
    const {onEvent} = this.props;
    if (typeof onEvent === 'function') {
      this.props.onEvent(name, data);
    }
  }

  @autobind
  handleMouseMove() {
    const {
      activityList
    } = this;
    if (activityList.isScrolledToBottom()) {
      this.acknowledgeLastActivity();
    }
  }

  /**
   * Flag or unflag activity by Id
   *
   * @param {String} activityId
   * @returns {undefined}
   */
  @autobind
  handleActivityFlag(activityId) {
    const {props} = this;
    const {
      activities,
      flags,
      sparkInstance
    } = props;
    const activity = Object.values(activities).find((act) => act.id === activityId);
    if (activity) {
      const foundFlag = flags.getIn(['flags', activity.url]);
      if (foundFlag && !foundFlag.get('isInFlight')) {
        props.removeFlagFromServer(foundFlag, sparkInstance);
      }
      else if (!foundFlag) {
        props.flagActivity(activity, sparkInstance);
      }
    }
  }

  /**
   * Displays modal confirming activity delete
   *
   * @param {String} activityId
   * @returns {undefined}
   */
  @autobind
  handleActivityDelete(activityId) {
    const {props} = this;
    props.confirmDeleteActivity(activityId);
  }

  /**
   * Does the actual deletion of the activity after confirmation modal
   *
   * @returns {undefined}
   */
  @autobind
  handleConfirmActivityDelete() {
    const {props} = this;
    const {
      activities,
      conversation,
      sparkInstance,
      widgetMessage
    } = props;

    const activityId = widgetMessage.get('deletingActivityId');

    const activity = Object.values(activities).find((act) => act.id === activityId);
    if (activity) {
      props.deleteActivityAndDismiss(conversation, activity, sparkInstance);
    }
    else {
      props.hideDeleteModal();
    }
  }


  /**
   * Dismisses the confirmation modal
   *
   * @returns {undefined}
   */
  @autobind
  handleCancelActivityDelete() {
    this.props.hideDeleteModal();
  }

  @autobind
  handleActivityRetry(activityId) {
    const {props} = this;
    const {
      activity,
      sparkInstance
    } = this.props;
    const failedActivity = activity.getIn(['activityFailures', activityId]);
    return props.retryFailedActivity(failedActivity, sparkInstance);
  }

  /**
   * Adds dropped file into activity store
   *
   * @param {array} acceptedFiles
   * @returns {undefined}
   */
  @autobind
  handleFileDrop(acceptedFiles) {
    const {props} = this;
    const {
      activity,
      conversation,
      sparkInstance
    } = props;
    const files = constructFiles(acceptedFiles);
    if (files.length) {
      props.addFiles(conversation, activity, files, sparkInstance);
    }
  }

  /**
   * Perform actions when activity list scrolls
   *
   * @returns {undefined}
   */
  @autobind
  handleScroll() {
    const {
      props,
      activityList
    } = this;
    const {
      conversationId,
      firstActivity,
      sparkInstance
    } = props;

    if (activityList.isScrolledToBottom()) {
      props.setScrolledUp(false);
      if (document.hasFocus()) {
        this.acknowledgeLastActivity();
      }
    }
    else {
      props.setScrolledUp(true);
    }

    if (activityList.isScrolledToTop() && firstActivity.verb !== 'create') {
      // Store scroll position before loading messages so the window
      // doesn't jump after they load
      props.setScrollPosition({scrollTop: activityList.getScrollTop()});
      props.loadPreviousMessages(
        conversationId,
        firstActivity.published,
        sparkInstance
      );
    }
  }

  /**
   * Renders the conversation area of the widget
   *
   * @returns {object}
   */
  render() {
    const {props} = this;
    const {
      conversation,
      conversationId,
      features,
      sparkInstance,
      currentUser,
      widgetMessage
    } = props;
    const {formatMessage} = this.props.intl;

    const scrollToBottomLabel = widgetMessage.get('hasNewMessage') ? formatMessage(messages.newMessagesMessage) : '';

    if (conversation && conversationId) {
      const isLoadingHistoryUp = conversation.getIn(['status', 'isLoadingHistoryUp']);
      const toUser = this.getToUserFromConversation(conversation);
      const displayName = conversation.get('displayName') || toUser && toUser.get('displayName');
      const showMentions = features.getIn(['items', FEATURE_MENTIONS]);

      return (
        <Dropzone
          activeClassName={styles.activeDropzone}
          className={styles.dropzone}
          disableClick
          disablePreview
          onDrop={this.handleFileDrop}
        >
          <div className={classNames('ciscospark-widget-message-main-area', styles.mainArea)} onMouseMove={this.handleMouseMove}>
            <div className={classNames('ciscospark-activity-list-wrapper', styles.activityListWrapper)}>
              <ScrollingActivity
                isLoadingHistoryUp={isLoadingHistoryUp}
                onScroll={this.handleScroll}
                ref={this.getActivityList}
              >
                {
                  currentUser &&
                  <ActivityList
                    newMessagesMessage={formatMessage(messages.newMessagesMessage)}
                    onActivityDelete={this.handleActivityDelete}
                    onActivityFlag={this.handleActivityFlag}
                    onActivityRetry={this.handleActivityRetry}
                  />
                }
                <div className={classNames('ciscospark-indicators', styles.indicators)}><ReadReceipts /></div>
              </ScrollingActivity>
              {
                widgetMessage.get('showScrollToBottomButton') &&
                <ScrollToBottomButton
                  label={scrollToBottomLabel}
                  onClick={this.handleScrollToBottom}
                />
              }
            </div>
            <div className={classNames('ciscospark-message-composer-wrapper', styles.messageComposerWrapper)}>
              <MessageComposer
                conversation={conversation}
                onSubmit={this.handleMessageSubmit}
                placeholder={formatMessage(messages.messageComposerPlaceholder, {displayName})}
                showMentions={showMentions}
                sparkInstance={sparkInstance}
              />
            </div>
          </div>
          {
            widgetMessage.get('showAlertModal') &&
            <ConfirmationModal
              actionButtonLabel={formatMessage(messages.deleteButtonLabel)}
              body={formatMessage(messages.deleteAlertBody)}
              cancelButtonLabel={formatMessage(messages.cancelButtonLabel)}
              onClickActionButton={this.handleConfirmActivityDelete}
              onClickCancelButton={this.handleCancelActivityDelete}
              title={formatMessage(messages.deleteAlertTitle)}
            />
          }
          <Notifications onEvent={this.handleEvent} />
          <Cover message={formatMessage(messages.dropzoneCoverMessage)} />
        </Dropzone>
      );
    }
    if (conversation && conversation.getIn(['status', 'error'])) {
      return (
        <ErrorDisplay
          secondaryTitle={conversation.getIn(['status', 'error', 'description'])}
          title={formatMessage(messages.errorConversation)}
        />
      );
    }
    return <LoadingScreen />;
  }
}

const injectedPropTypes = {
  avatar: PropTypes.object.isRequired,
  activity: PropTypes.object.isRequired,
  conversation: PropTypes.object.isRequired,
  flags: PropTypes.object.isRequired,
  sparkInstance: PropTypes.object,
  user: PropTypes.object.isRequired,
  widgetMessage: PropTypes.object.isRequired,
  acknowledgeActivityOnServer: PropTypes.func.isRequired,
  addFiles: PropTypes.func.isRequired,
  confirmDeleteActivity: PropTypes.func.isRequired,
  createConversation: PropTypes.func.isRequired,
  createNotification: PropTypes.func.isRequired,
  deleteActivityAndDismiss: PropTypes.func.isRequired,
  fetchAvatarsForUsers: PropTypes.func.isRequired,
  fetchFlags: PropTypes.func.isRequired,
  flagActivity: PropTypes.func.isRequired,
  getConversation: PropTypes.func.isRequired,
  hideDeleteModal: PropTypes.func.isRequired,
  loadMissingActivities: PropTypes.func.isRequired,
  loadPreviousMessages: PropTypes.func.isRequired,
  removeFlagFromServer: PropTypes.func.isRequired,
  retryFailedActivity: PropTypes.func.isRequired,
  removeInflightActivity: PropTypes.func.isRequired,
  setScrollPosition: PropTypes.func.isRequired,
  setScrolledUp: PropTypes.func.isRequired,
  setTyping: PropTypes.func.isRequired,
  showScrollToBottomButton: PropTypes.func.isRequired,
  updateHasNewMessage: PropTypes.func.isRequired,
  updateWidgetState: PropTypes.func.isRequired,
  space: PropTypes.object.isRequired,
  features: PropTypes.object.isRequired,
  getFeature: PropTypes.func.isRequired,
  participants: PropTypes.array.isRequired,
  activities: PropTypes.object.isRequired,
  conversationId: PropTypes.string
};

export const ownPropTypes = {
  hasAcknowledgementsDisabled: PropTypes.bool,
  muteNotifications: PropTypes.bool,
  onEvent: PropTypes.func,
  spaceId: PropTypes.string,
  toPersonEmail: PropTypes.string,
  toPersonId: PropTypes.string,
  eventNames: PropTypes.object
};

MessageWidget.propTypes = {
  ...ownPropTypes,
  ...injectedPropTypes
};

const ConnectedMessageWidget = wrapConversationMercury(MessageWidget);

export default connect(
  getMessageWidgetProps,
  (dispatch) => bindActionCreators({
    acknowledgeActivityOnServer,
    addFiles,
    confirmDeleteActivity,
    createConversation,
    createNotification,
    deleteActivityAndDismiss,
    fetchAvatarsForUsers,
    fetchFlags,
    flagActivity,
    getConversation,
    hideDeleteModal,
    loadMissingActivities,
    loadPreviousMessages,
    removeFlagFromServer,
    removeInflightActivity,
    retryFailedActivity,
    setScrollPosition,
    setScrolledUp,
    setTyping,
    showScrollToBottomButton,
    subscribeToPresenceUpdates,
    unsubscribeFromPresenceUpdates,
    updateHasNewMessage,
    updateWidgetState,
    getFeature
  }, dispatch)
)(ConnectedMessageWidget);
