// @ciscospark/widget-message/src/reducer.js
import {fromJS} from 'immutable';
import {combineReducers} from 'redux';

import conversation from '@ciscospark/redux-module-conversation';
import activity from '@ciscospark/redux-module-activity';
import avatar from '@ciscospark/redux-module-avatar';
import flags from '@ciscospark/redux-module-flags';
import indicators from '@ciscospark/redux-module-indicators';
import {reducer as notifications} from '@ciscospark/react-container-notifications';
import {reducer as messageComposer} from '@ciscospark/react-container-message-composer';
import share from '@ciscospark/redux-module-share';
import user from '@ciscospark/redux-module-user';
import spark from '@ciscospark/react-redux-spark';

import {
  UPDATE_WIDGET_STATE,
  SET_SCROLL_POSITION
} from './actions';

export const initialState = fromJS({
  deletingActivityId: null,
  isListeningToActivity: false,
  isListeningToTyping: false,
  showAlertModal: false,
  showScrollToBottomButton: false,
  hasNewMessage: false,
  scrollPosition: {}
});

export function reducer(state = initialState, action) {
  switch (action.type) {
  case UPDATE_WIDGET_STATE:
    return state.merge(action.payload.state);

  case SET_SCROLL_POSITION:
    return state.set(`scrollPosition`, action.payload.scrollPosition);

  default:
    return state;
  }
}

export const reducers = {
  activity,
  avatar,
  conversation,
  flags,
  indicators,
  notifications,
  share,
  user,
  spark,
  messageComposer,
  widgetMessage: reducer
};

export default combineReducers(reducers);
