// @ciscospark/react-container-scrollable/src/index.js
import React, {Component, PropTypes} from 'react';

import styles from './styles.css';

class ScrollableComponent extends Component {
  constructor(props) {
    super(props);
    this.getNode = this.getNode.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    return nextProps !== this.props;
  }

  getNode(node) {
    this.node = node;
  }

  getScrollHeight() {
    return this.node.scrollHeight;
  }

  getScrollTop() {
    return this.node.scrollTop;
  }

  setScrollTop(top) {
    this.node.scrollTop = top;
  }

  scrollToBottom() {
    const node = this.node;
    node.scrollTop = node.scrollHeight;
  }

  isScrolledToTop() {
    return this.node.scrollTop < 100;
  }

  isScrolledToBottom() {
    const node = this.node;
    return node.scrollHeight - node.offsetHeight - node.scrollTop < 150;
  }

  render() {
    return (
      <div className={styles.scrollable} onScroll={this.props.onScroll} ref={this.getNode}>
        {this.props.children}
      </div>
    );
  }
}

ScrollableComponent.propTypes = {
  children: PropTypes.element.isRequired,
  onScroll: PropTypes.func
};

export default ScrollableComponent;
