// @ciscospark/react-component-textarea/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import styles from './styles.css';

export default function TextArea(props) {
  const {
    textAreaClassName,
    onBlur,
    onChange,
    onFocus,
    onKeyDown,
    onSubmit,
    placeholder,
    rows,
    value
  } = props;

  const textAreaProps = {
    className: classNames(`ciscospark-textarea`, styles.textarea, textAreaClassName),
    onBlur,
    onChange,
    onFocus,
    onKeyDown,
    onSubmit,
    placeholder,
    rows
  };

  // Only set text area value if onChange method exists, otherwise use existing value
  if (onChange && value !== undefined) {
    textAreaProps.value = value;
  }

  return (
    <textarea {...textAreaProps} />
  );
}

TextArea.propTypes = {
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onKeyDown: PropTypes.func,
  onSubmit: PropTypes.func,
  placeholder: PropTypes.string,
  rows: PropTypes.number,
  textAreaClassName: PropTypes.string,
  value: PropTypes.string
};
