// @ciscospark/react-component-presence-avatar/src/index.js
import React from 'react';
import PropTypes from 'prop-types';

import {Avatar} from '@collab-ui/react';

import {
  PRESENCE_TYPE_ACTIVE,
  PRESENCE_TYPE_INACTIVE,
  PRESENCE_TYPE_DND,
  PRESENCE_TYPE_OOO
} from '@ciscospark/redux-module-presence';

const propTypes = {
  image: PropTypes.string,
  isSelfAvatar: PropTypes.bool,
  isTyping: PropTypes.bool,
  name: PropTypes.string,
  size: PropTypes.string,
  spaceType: PropTypes.string,
  presenceStatus: PropTypes.string
};

const defaultProps = {
  image: undefined,
  isSelfAvatar: false,
  isTyping: false,
  name: undefined,
  size: '40',
  spaceType: undefined,
  presenceStatus: undefined
};

function PresenceAvatar(props) {
  const {
    image,
    isSelfAvatar,
    isTyping,
    name,
    size,
    spaceType,
    presenceStatus
  } = props;
  let type;

  // Do not show presence on self avatars
  if (!isSelfAvatar) {
    switch (presenceStatus) {
      case PRESENCE_TYPE_ACTIVE:
        type = 'active';
        break;
      case PRESENCE_TYPE_INACTIVE:
        type = 'inactive';
        break;
      case PRESENCE_TYPE_OOO:
        type = 'ooo';
        break;
      case PRESENCE_TYPE_DND:
        type = 'dnd';
        break;
      default:
        type = 'default';
    }
    if (isTyping) {
      type = 'typing';
    }
    if (spaceType === 'group') {
      type = 'group';
    }
  }
  else {
    type = 'self';
  }

  return (
    <div>
      <Avatar
        title={name}
        type={type}
        src={image}
        size={size}
      />
    </div>
  );
}

PresenceAvatar.propTypes = propTypes;
PresenceAvatar.defaultProps = defaultProps;

export default PresenceAvatar;
