// @ciscospark/react-component-button/src/index.js
import React, {PropTypes} from 'react';
import classNames from 'classnames';

import Icon from '@ciscospark/react-component-icon';

import styles from './styles.css';

export default function Button(props) {
  const {
    buttonClassName,
    children,
    label,
    labelPosition,
    onClick,
    title,
    iconType
  } = props;

  return (
    <div className={classNames(`ciscospark-button-container`, styles.buttonContainer)}>
      <button
        aria-label={label}
        className={classNames(`ciscospark-button`, styles.button, buttonClassName)}
        onClick={onClick} onKeyUp={onClick}
      >
        {
          iconType &&
          <Icon title={title} type={iconType} />
        }
        {
          labelPosition !== `bottom` &&
          label
        }
        {children}
      </button>
      {
        label && labelPosition === `bottom` &&
        <div className={classNames(`ciscospark-label`, styles.label)}>{label}</div>
      }
    </div>
  );
}

Button.propTypes = {
  buttonClassName: PropTypes.string,
  children: PropTypes.arrayOf(PropTypes.element),
  iconType: PropTypes.string,
  label: PropTypes.string,
  labelPosition: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  title: PropTypes.string
};
